// zent/es/colorpicker/SketchPresetColors.js
import { __assign } from "tslib";
import * as React from 'react';
import reactCSS from './helpers/reactcss';
import { Swatch } from './common';
var SketchPresetColors = function (_a) {
    var colors = _a.colors, onClick = _a.onClick, prefix = _a.prefix, type = _a.type;
    var styles = reactCSS({
        default: {
            colors: {
                margin: '0 -10px',
                padding: '10px 0 0 10px',
                borderTop: '1px solid #eee',
                display: 'flex',
                flexWrap: 'wrap',
                position: 'relative',
            },
            swatchWrap: {
                width: '16px',
                height: '16px',
                margin: '0 10px 10px 0',
            },
            swatch: {
                borderRadius: '3px',
                boxShadow: 'inset 0 0 0 1px rgba(0,0,0,.15)',
            },
        },
        'no-presets': {
            colors: {
                display: 'none',
            },
        },
    }, {
        'no-presets': !colors || !colors.length,
    });
    var handleClick = function (hex, e) {
        onClick({
            hex: hex,
            source: 'hex',
        }, e);
    };
    if (type === 'simple') {
        return (React.createElement("div", { className: prefix + "-colorpicker-colors-select", "data-zv": '7.4.3' }, colors.map(function (color) { return (React.createElement("div", { key: color, className: prefix + "-colorpicker-colors-select__preview", style: { backgroundColor: color }, onClick: function () { return onClick(color); }, title: color, "data-zv": '7.4.3' })); })));
    }
    return (React.createElement("div", { style: styles.colors, className: prefix + "-colorpicker-colors", "data-zv": '7.4.3' }, colors.map(function (colorObjOrString) {
        var c = typeof colorObjOrString === 'string'
            ? { color: colorObjOrString }
            : colorObjOrString;
        return (React.createElement("div", { key: c.color, style: styles.swatchWrap, "data-zv": '7.4.3' },
            React.createElement(Swatch, __assign({}, c, { style: styles.swatch, onClick: handleClick }))));
    })));
};
export default SketchPresetColors;
