// uxcore-alert/build/Alert.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames4 = require('classnames');

var _classnames5 = _interopRequireDefault(_classnames4);

var _uxcoreIcon = require('uxcore-icon');

var _uxcoreIcon2 = _interopRequireDefault(_uxcoreIcon);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Alert Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author eternaslky
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2018, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var Alert = function (_React$Component) {
  _inherits(Alert, _React$Component);

  function Alert(props) {
    _classCallCheck(this, Alert);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      closed: false
    };
    return _this;
  }

  Alert.prototype.toggleShow = function toggleShow() {
    this.setState({
      closed: !this.state.closed
    });
  };

  Alert.prototype.handleClose = function handleClose(e) {
    this.setState({
      closed: true
    });
    this.props.onClose.call(this, e);
  };

  Alert.prototype.renderContent = function renderContent() {
    var _classnames;

    var me = this;
    var iconMap = {
      message: 'uxcore-icon uxicon-tishi-full',
      error: 'uxcore-icon uxicon-biaodanlei-tongyongqingchu',
      warning: 'uxcore-icon uxicon-jinggao-full',
      success: 'uxcore-icon uxicon-chenggong-full',
      question: 'uxcore-icon uxicon-yiwen-full',
      stop: 'kuma-icon kuma-icon-prohibition',
      wait: 'kuma-icon kuma-icon-wait'
    };
    var _me$props = me.props,
        prefixCls = _me$props.prefixCls,
        iconPrefixCls = _me$props.iconPrefixCls,
        type = _me$props.type,
        closable = _me$props.closable,
        description = _me$props.description,
        hasContainer = _me$props.hasContainer,
        className = _me$props.className;

    var iconClassName = iconMap[type];
    // 保证 className 加在最外层容器上
    return _react2["default"].createElement(
      'p',
      {
        className: (0, _classnames5["default"])((_classnames = {}, _defineProperty(_classnames, prefixCls + ' ' + prefixCls + '-' + type, true), _defineProperty(_classnames, className, !!className && !closable && !description && !hasContainer), _classnames))
      },
      _react2["default"].createElement('i', {
        className: (0, _classnames5["default"])(prefixCls + '-icon ' + iconClassName, _defineProperty({}, iconPrefixCls + ' ' + iconPrefixCls + '-' + type, ['kuma-icon', 'uxcore-icon'].indexOf(iconPrefixCls) === -1))
      }),
      _react2["default"].createElement(
        'span',
        null,
        me.props.message
      )
    );
  };

  Alert.prototype.renderDes = function renderDes() {
    var me = this;
    var _me$props2 = me.props,
        description = _me$props2.description,
        prefixCls = _me$props2.prefixCls;

    if (description) {
      return _react2["default"].createElement(
        'p',
        { className: prefixCls + ' ' + prefixCls + '-follow' },
        description
      );
    }
    return null;
  };

  Alert.prototype.renderClose = function renderClose() {
    var _this2 = this;

    var me = this;
    var prefixCls = me.props.prefixCls;

    if (me.props.closable) {
      if (me.props.closeText) {
        return _react2["default"].createElement(
          'span',
          {
            className: prefixCls + '-close',
            onClick: me.handleClose.bind(me)
          },
          me.props.closeText
        );
      }
      return _react2["default"].createElement(_uxcoreIcon2["default"], {
        name: 'biaoqian-qingchu',
        className: prefixCls + '-close',
        onClick: function onClick(e) {
          _this2.handleClose(e);
        }
      });
    }
    return null;
  };

  Alert.prototype.render = function render() {
    var me = this;
    var _me$props3 = me.props,
        prefixCls = _me$props3.prefixCls,
        type = _me$props3.type,
        closable = _me$props3.closable,
        hasContainer = _me$props3.hasContainer,
        description = _me$props3.description,
        className = _me$props3.className,
        size = _me$props3.size;

    var html = void 0;
    var content = me.renderContent();
    if (hasContainer || closable || !!description) {
      var _classnames3;

      html = _react2["default"].createElement(
        'div',
        {
          className: (0, _classnames5["default"])((_classnames3 = {}, _defineProperty(_classnames3, prefixCls + '-container ' + prefixCls + '-container-' + type, true), _defineProperty(_classnames3, prefixCls + '-container-closable', closable), _defineProperty(_classnames3, className, !!className), _defineProperty(_classnames3, prefixCls + '-size-large', size === 'large'), _classnames3))
        },
        content,
        me.renderDes(),
        me.renderClose()
      );
    } else {
      html = content;
    }
    return this.state.closed ? null : html;
  };

  return Alert;
}(_react2["default"].Component);

Alert.defaultProps = {
  prefixCls: 'kuma-alert',
  iconPrefixCls: 'kuma-icon',
  type: 'message',
  onClose: function onClose() {},
  hasContainer: false,
  size: 'normal'
};

// http://facebook.github.io/react/docs/reusable-components.html
Alert.propTypes = {
  prefixCls: _propTypes2["default"].string,
  iconPrefixCls: _propTypes2["default"].string,
  type: _propTypes2["default"].oneOf(['message', 'error', 'warning', 'success', 'question', 'stop', 'wait']),
  closable: _propTypes2["default"].bool,
  hasContainer: _propTypes2["default"].bool,
  closeText: _propTypes2["default"].oneOfType([_propTypes2["default"].element, _propTypes2["default"].string]),
  message: _propTypes2["default"].oneOfType([_propTypes2["default"].element, _propTypes2["default"].string]),
  description: _propTypes2["default"].oneOfType([_propTypes2["default"].element, _propTypes2["default"].string]),
  onClose: _propTypes2["default"].func,
  size: _propTypes2["default"].oneOf(['normal', 'large'])

};

Alert.displayName = 'Alert';

exports["default"] = Alert;
module.exports = exports['default'];
