// uxcore-popover/build/Popover.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _rcTooltip = require('rc-tooltip');

var _rcTooltip2 = _interopRequireDefault(_rcTooltip);

var _uxcoreButton = require('uxcore-button');

var _uxcoreButton2 = _interopRequireDefault(_uxcoreButton);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactLifecyclesCompat = require('react-lifecycles-compat');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Popover Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author eternalsky
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2015, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var Popover = function (_React$Component) {
  _inherits(Popover, _React$Component);

  function Popover(props) {
    _classCallCheck(this, Popover);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      visible: 'visible' in props ? props.visible : false
    };
    return _this;
  }

  Popover.getDerivedStateFromProps = function getDerivedStateFromProps(props) {
    var states = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    if ('visible' in props && props.visible !== states._visible) {
      return {
        visible: props.visible,
        _visible: props.visible
      };
    }
    return null;
  };

  Popover.prototype.handleOkClick = function handleOkClick() {
    var me = this;
    me.props.onOk(function () {
      if (!('visible' in me.props)) {
        me.setState({
          visible: false
        });
      }
    });
  };

  Popover.prototype.handleCancelClick = function handleCancelClick() {
    var me = this;

    if (!('visible' in me.props)) {
      me.setState({
        visible: false
      }, function () {
        me.props.onCancel();
      });
    } else {
      me.props.onCancel();
    }
  };

  Popover.prototype.handleVisibleChange = function handleVisibleChange(visible) {
    var me = this;
    if (!('visible' in me.props)) {
      me.setState({
        visible: visible
      }, function () {
        me.props.onVisibleChange(visible);
      });
    } else {
      me.props.onVisibleChange(visible);
    }
  };

  Popover.prototype.generateOverlay = function generateOverlay() {
    var me = this;
    var arr = [];
    if (me.props.title) {
      arr.push(_react2["default"].createElement(
        'div',
        { key: 'title', className: me.props.prefixCls + '-title' },
        me.props.title
      ));
    }
    if (me.props.overlay) {
      arr.push(_react2["default"].createElement(
        'div',
        { key: 'content', className: me.props.prefixCls + '-content' },
        me.props.overlay,
        me.renderButton()
      ));
    }
    return _react2["default"].createElement(
      'div',
      null,
      arr
    );
  };

  Popover.prototype.renderButton = function renderButton() {
    var me = this;
    if (me.props.showButton) {
      return _react2["default"].createElement(
        'div',
        { className: me.props.prefixCls + '-button-group' },
        _react2["default"].createElement(
          _uxcoreButton2["default"],
          {
            size: 'small', type: 'secondary', onClick: me.handleCancelClick.bind(me)
          },
          me.props.cancelText
        ),
        _react2["default"].createElement(
          _uxcoreButton2["default"],
          {
            size: 'small', type: 'primary', onClick: me.handleOkClick.bind(me)
          },
          me.props.okText
        )
      );
    }
    return null;
  };

  Popover.prototype.render = function render() {
    var me = this;

    return _react2["default"].createElement(
      _rcTooltip2["default"],
      {
        placement: me.props.placement,
        arrowContent: me.props.arrowContent,
        overlayClassName: me.props.overlayClassName,
        prefixCls: me.props.prefixCls,
        align: me.props.align,
        visible: me.state.visible,
        onVisibleChange: me.handleVisibleChange.bind(me),
        mouseEnterDelay: me.props.delay,
        transitionName: me.props.transitionName,
        trigger: me.props.trigger,
        overlay: me.generateOverlay(),
        getTooltipContainer: me.props.getTooltipContainer
      },
      me.props.children
    );
  };

  return Popover;
}(_react2["default"].Component);

Popover.defaultProps = {
  prefixCls: 'kuma-popover',
  transitionName: 'popover-slide',
  delay: 0,
  placement: 'top',
  trigger: 'hover',
  onOk: function onOk(cb) {
    cb();
  },
  onCancel: function onCancel() {},
  okText: '确定',
  cancelText: '取消',
  showButton: false,
  arrowContent: _react2["default"].createElement('div', { className: 'kuma-popover-arrow-inner' }),
  onVisibleChange: function onVisibleChange() {}
};

// http://facebook.github.io/react/docs/reusable-components.html
Popover.propTypes = {
  prefixCls: _propTypes2["default"].string,
  delay: _propTypes2["default"].number,
  overlayClassName: _propTypes2["default"].string,
  placement: _propTypes2["default"].oneOf(['top', 'bottom', 'left', 'right', 'topLeft', 'bottomLeft', 'leftTop', 'rightTop', 'topRight', 'bottomRight', 'leftBottom', 'rightBottom']),
  trigger: _propTypes2["default"].oneOf(['hover', 'click']),
  onOk: _propTypes2["default"].func,
  onCancel: _propTypes2["default"].func,
  okText: _propTypes2["default"].string,
  cancelText: _propTypes2["default"].string,
  visible: _propTypes2["default"].bool,
  showButton: _propTypes2["default"].bool,
  onVisibleChange: _propTypes2["default"].func
};

Popover.displayName = 'Popover';

(0, _reactLifecyclesCompat.polyfill)(Popover);

exports["default"] = Popover;
module.exports = exports['default'];
