// uxcore-cascade-select/build/Search.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Search = function Search(_ref) {
  var text = _ref.text,
      disabled = _ref.disabled,
      placeholder = _ref.placeholder,
      searchOption = _ref.searchOption,
      value = _ref.value,
      onValueChange = _ref.onValueChange,
      onSearchResultChange = _ref.onSearchResultChange,
      onInputFocus = _ref.onInputFocus,
      onInputBlur = _ref.onInputBlur;
  return _react2["default"].createElement('input', {
    type: 'text',
    placeholder: placeholder,
    style: { width: '100%', border: 'none', background: 'transparent' },
    disabled: disabled,
    value: value !== null ? value : text,
    onChange: function onChange(e) {
      var keywords = e.target.value;
      // 修复IE11下bug
      if (value === null && text === '' && keywords === '') {
        onValueChange(null);
      } else {
        onValueChange(keywords);
      }
      if (searchOption) {
        searchOption.doSearch(keywords, function (searchResult) {
          onSearchResultChange(searchResult);
        });
      }
    },
    onFocus: onInputFocus,
    onBlur: onInputBlur
  });
};

Search.propTypes = {
  value: _propTypes2["default"].string,
  text: _propTypes2["default"].string,
  disabled: _propTypes2["default"].bool,
  placeholder: _propTypes2["default"].string,
  searchOption: _propTypes2["default"].object,
  onSearchResultChange: _propTypes2["default"].func,
  onValueChange: _propTypes2["default"].func,
  onInputFocus: _propTypes2["default"].func,
  onInputBlur: _propTypes2["default"].func
};

Search.renderResult = function (result, onSelect) {
  return _react2["default"].createElement(
    'div',
    {
      className: '',
      style: {
        background: '#fff',
        border: '1px solid #d9d9d9',
        boxShadow: '0 1px 4px 0 rgba(31,56,88,.15)',
        borderRadius: '3px',
        minWidth: '200px'
      }
    },
    _react2["default"].createElement(
      'ul',
      { className: 'kuma-select2-dropdown-menu' },
      result.map(function (item) {
        return _react2["default"].createElement(
          'li',
          {
            key: item.value,
            className: 'kuma-select2-dropdown-menu-item',
            onClick: function onClick() {
              return onSelect(item);
            },
            onMouseEnter: function onMouseEnter(e) {
              e.target.className += ' kuma-select2-dropdown-menu-item-active';
            },
            onMouseLeave: function onMouseLeave(e) {
              e.target.className = e.target.className.replace(' kuma-select2-dropdown-menu-item-active', '');
            }
          },
          item.label
        );
      })
    )
  );
};

exports["default"] = Search;
module.exports = exports['default'];
