// zent/es/mention/position.js
import Popover from '../popover';
export var getPopoverBottomPosition = function (instance) {
    return Popover.Position.create(function (anchorBoundingBox, containerBoundingBox, contentDimension, options) {
        return {
            getCSSStyle: function () {
                var left = anchorBoundingBox.left, top = anchorBoundingBox.top, right = anchorBoundingBox.right, bottom = anchorBoundingBox.bottom;
                var position = instance.state.position;
                var x = left + position.left;
                var y = top + options.cushion + position.top + position.height;
                var inputStyles = getComputedStyle(instance.input);
                var leftSpace = parseInt(inputStyles.paddingLeft, 10) +
                    parseInt(inputStyles.borderLeftWidth, 10);
                var rightSpace = parseInt(inputStyles.paddingRight, 10) +
                    parseInt(inputStyles.borderRightWidth, 10);
                if (x > right - rightSpace) {
                    x = right - rightSpace;
                }
                if (x < left + leftSpace) {
                    x = left + leftSpace;
                }
                if (y < top) {
                    y = top;
                }
                if (y > bottom) {
                    y = bottom;
                }
                return {
                    position: 'absolute',
                    left: Math.round(x) + "px",
                    top: Math.round(y) + "px",
                };
            },
            name: 'position-mention-bottom-left',
        };
    });
};
export var getPopoverTopPosition = function (instance) {
    return Popover.Position.create(function (anchorBoundingBox, containerBoundingBox, contentDimension, options) {
        return {
            getCSSStyle: function () {
                var left = anchorBoundingBox.left, top = anchorBoundingBox.top, right = anchorBoundingBox.right, bottom = anchorBoundingBox.bottom;
                var contentHeight = contentDimension.height;
                var position = instance.state.position;
                var x = left + position.left;
                var y = top - contentHeight - options.cushion + position.top;
                var inputStyles = getComputedStyle(instance.input);
                var leftSpace = parseInt(inputStyles.paddingLeft, 10) +
                    parseInt(inputStyles.borderLeftWidth, 10);
                var rightSpace = parseInt(inputStyles.paddingRight, 10) +
                    parseInt(inputStyles.borderRightWidth, 10);
                if (x > right - rightSpace) {
                    x = right - rightSpace;
                }
                if (x < left + leftSpace) {
                    x = left + leftSpace;
                }
                if (y + contentHeight < top) {
                    y = top - contentHeight;
                }
                if (y + contentHeight > bottom) {
                    y = bottom - contentHeight;
                }
                return {
                    position: 'absolute',
                    left: Math.round(x) + "px",
                    top: Math.round(y) + "px",
                };
            },
            name: 'position-mention-top-left',
        };
    });
};
