// zent/es/placeholder/presets/RichTextBlock.js
import { __assign, __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import cx from 'classnames';
import Circle from '../shapes/Circle';
import Rectangle from '../shapes/Rectangle';
import TextBlock from './TextBlock';
var RichTextBlock = (function (_super) {
    __extends(RichTextBlock, _super);
    function RichTextBlock() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RichTextBlock.prototype.render = function () {
        var _a = this.props, className = _a.className, style = _a.style, rows = _a.rows, shape = _a.shape, prefix = _a.prefix, size = _a.size, animate = _a.animate, dashed = _a.dashed, widths = _a.widths, dashSegments = _a.dashSegments, lineSpacing = _a.lineSpacing;
        var classes = cx(prefix + "-placeholder-richtext-block", className);
        var shapeStyle = { marginRight: 10 };
        return (React.createElement("div", { className: classes, style: __assign(__assign({}, style), { display: 'flex' }), "data-zv": '7.4.3' },
            shape === 'circle' ? (React.createElement(Circle, { style: shapeStyle, diameter: size, animate: animate })) : (React.createElement(Rectangle, { style: shapeStyle, width: size, height: size, animate: animate })),
            React.createElement(TextBlock, { rows: rows, animate: animate, dashed: dashed, widths: widths, dashSegments: dashSegments, lineSpacing: lineSpacing })));
    };
    RichTextBlock.defaultProps = {
        style: {},
        className: '',
        prefix: 'zent',
        shape: 'circle',
        animate: true,
        dashed: true,
        size: 80,
    };
    return RichTextBlock;
}(PureComponent));
export default RichTextBlock;
