// uxcore-load-more/build/LoadMore.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; /**
                                                                                                                                                                                                                                                                               * LoadMore Component for uxcore
                                                                                                                                                                                                                                                                               * @author peijie.dpj
                                                                                                                                                                                                                                                                               *
                                                                                                                                                                                                                                                                               * Copyright 2015-2017, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                               * All rights reserved.
                                                                                                                                                                                                                                                                               */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _i18n = require('./i18n');

var _i18n2 = _interopRequireDefault(_i18n);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var scrollTimer = void 0;

function isElementInView(element) {
  if (!element || (typeof window === 'undefined' ? 'undefined' : _typeof(window)) !== 'object') {
    return false;
  }

  var threshold = 0;
  var innerHeight = window.innerHeight || document.documentElement.clientHeight || document.body.clientHeight;
  var viewTop = window.scrollY || window.pageYOffset || document.documentElement.scrollTop;
  var viewBottom = viewTop + innerHeight;
  var elementTop = !!element && !!element.getBoundingClientRect && element.getBoundingClientRect().top + viewTop;
  var elementBottom = elementTop + element.offsetHeight;
  var offset = threshold / 100 * innerHeight;

  return viewTop - offset <= elementBottom && elementTop <= viewBottom + offset;
}

var LoadMore = function (_React$Component) {
  _inherits(LoadMore, _React$Component);

  function LoadMore(props) {
    _classCallCheck(this, LoadMore);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.lang = _i18n2["default"][props.locale.toLocaleLowerCase()];
    return _this;
  }

  LoadMore.prototype.componentDidMount = function componentDidMount() {
    var me = this;
    var props = me.props;

    var hasView = props.trigger.some(function (item) {
      return item === 'view';
    });

    if (hasView && (typeof window === 'undefined' ? 'undefined' : _typeof(window)) === 'object') {
      if (window.addEventListener) {
        window.addEventListener('scroll', me.onScroll.bind(me));
      } else if (window.attachEvent) {
        window.attachEvent('onscroll', me.onScroll.bind(me));
      }

      // 第一屏判断
      if (isElementInView(me.refs.viewMore)) {
        props.onLoadMore();
      }
    }
  };

  /**
   * 滚动事件
   */


  LoadMore.prototype.onScroll = function onScroll() {
    var me = this;
    var props = me.props;

    if (scrollTimer) {
      clearTimeout(scrollTimer);
    }

    scrollTimer = setTimeout(me.onScrollStop.bind(me), props.viewLoadDelay);
  };

  /**
   * 滚动停止一段时间之后才判断
   */


  LoadMore.prototype.onScrollStop = function onScrollStop() {
    var me = this;
    var props = me.props;

    if (isElementInView(me.refs.viewMore)) {
      props.onLoadMore();
    }
  };

  LoadMore.prototype.onClick = function onClick() {
    var me = this;
    var props = me.props;

    var hasClick = props.trigger.some(function (item) {
      return item === 'click';
    });

    if (hasClick) {
      props.onLoadMore();
    }
  };

  LoadMore.prototype.renderLoadStatus = function renderLoadStatus() {
    var me = this;
    var props = me.props;
    var lang = me.lang;
    var res = void 0;

    switch (props.status) {
      case 'loaded':
        res = _react2["default"].createElement(
          'a',
          {
            href: 'javascript:void(0)',
            className: props.prefixCls + '-status',
            ref: 'viewMore',
            onClick: me.onClick.bind(me)
          },
          'loadText' in props ? props.loadText : lang.viewMore
        );
        break;

      case 'loading':
        res = _react2["default"].createElement(
          'a',
          {
            href: 'javascript:void(0)',
            className: props.prefixCls + '-status'
          },
          _react2["default"].createElement('i', { className: props.prefixCls + '-icon-loading' }),
          'loadingText' in props ? props.loadingText : lang.loading
        );
        break;

      case 'noMore':
        res = _react2["default"].createElement(
          'a',
          {
            href: 'javascript:void(0)',
            className: props.prefixCls + '-status ' + props.prefixCls + '-noMore'
          },
          'noMoreText' in props ? props.noMoreText : lang.noMore
        );
        break;

      default:
        break;
    }

    return res;
  };

  LoadMore.prototype.render = function render() {
    var props = this.props;

    return _react2["default"].createElement(
      'div',
      {
        className: (0, _classnames3["default"])(props.prefixCls, _defineProperty({}, props.className, !!props.className))
      },
      this.renderLoadStatus()
    );
  };

  return LoadMore;
}(_react2["default"].Component);

LoadMore.defaultProps = {
  prefixCls: 'kuma-load-more',
  status: 'loaded',
  className: '',
  trigger: ['view', 'click'],
  onLoadMore: function onLoadMore() {},
  locale: 'zh-cn',
  viewLoadDelay: 150
};

// http://facebook.github.io/react/docs/reusable-components.html
LoadMore.propTypes = {
  prefixCls: _propTypes2["default"].string,
  status: _propTypes2["default"].oneOf(['loaded', 'loading', 'noMore']),
  className: _propTypes2["default"].string,
  trigger: _propTypes2["default"].array,
  onLoadMore: _propTypes2["default"].func,
  locale: _propTypes2["default"].oneOf(['zh-cn', 'en-us']),
  viewLoadDelay: _propTypes2["default"].number
};

LoadMore.displayName = 'LoadMore';

module.exports = LoadMore;
