// zent/es/pagination/LitePagination.js
import { __extends } from "tslib";
import layoutLite from './layout/lite';
import AbstractPagination from './impl/AbstractPagination';
var LitePagination = (function (_super) {
    __extends(LitePagination, _super);
    function LitePagination() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(LitePagination.prototype, "name", {
        get: function () {
            return 'lite';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LitePagination.prototype, "layoutFn", {
        get: function () {
            return layoutLite;
        },
        enumerable: true,
        configurable: true
    });
    LitePagination.defaultProps = {
        current: 1,
        pageSize: 10,
        buttonBordered: false,
        showQuickJumper: false,
        showSizeChanger: false,
    };
    return LitePagination;
}(AbstractPagination));
export { LitePagination };
export default LitePagination;
