// zent/es/colorpicker/common/Hue.js
import { __assign, __extends } from "tslib";
import * as React from 'react';
import { Component, createRef } from 'react';
import reactCSS from '../helpers/reactcss';
import * as hue from '../helpers/hue';
var Hue = (function (_super) {
    __extends(Hue, _super);
    function Hue() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.containerRef = createRef();
        _this.handleChange = function (e, skip) {
            var change = hue.calculateChange(e, skip, _this.props, _this.containerRef.current);
            change && _this.props.onChange(change, e);
        };
        _this.handleMouseDown = function (e) {
            _this.handleChange(e, true);
            window.addEventListener('mousemove', _this.handleChange);
            window.addEventListener('mouseup', _this.handleMouseUp);
        };
        _this.handleMouseUp = function () {
            _this.unbindEventListeners();
        };
        return _this;
    }
    Hue.prototype.componentWillUnmount = function () {
        this.unbindEventListeners();
    };
    Hue.prototype.unbindEventListeners = function () {
        window.removeEventListener('mousemove', this.handleChange);
        window.removeEventListener('mouseup', this.handleMouseUp);
    };
    Hue.prototype.render = function () {
        var styles = reactCSS({
            default: {
                hue: {
                    absolute: '0px 0px 0px 0px',
                    background: "linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%,\n            #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)",
                    borderRadius: this.props.radius,
                    boxShadow: this.props.shadow,
                },
                container: {
                    margin: '0 2px',
                    position: 'relative',
                    height: '100%',
                },
                pointer: {
                    position: 'absolute',
                    left: (this.props.hsl.h * 100) / 360 + "%",
                },
                slider: {
                    marginTop: '1px',
                    width: '4px',
                    borderRadius: '1px',
                    height: '8px',
                    boxShadow: '0 0 2px rgba(0, 0, 0, .6)',
                    background: '#fff',
                    transform: 'translateX(-2px)',
                },
            },
            vertical: {
                hue: {
                    background: "linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,\n            #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)",
                },
                pointer: {
                    left: '0px',
                    top: -((this.props.hsl.h * 100) / 360) + 100 + "%",
                },
            },
        }, { vertical: this.props.direction === 'vertical' });
        return (React.createElement("div", { style: styles.hue, className: "hue-area", "data-zv": '7.4.3' },
            React.createElement("div", { className: "hue-bar", style: styles.container, ref: this.containerRef, onMouseDown: this.handleMouseDown, onTouchMove: this.handleChange, onTouchStart: this.handleChange, "data-zv": '7.4.3' },
                React.createElement("div", { style: styles.pointer, "data-zv": '7.4.3' }, this.props.pointer ? (React.createElement(this.props.pointer, __assign({}, this.props))) : (React.createElement("div", { style: styles.slider, "data-zv": '7.4.3' }))))));
    };
    return Hue;
}(Component));
export default Hue;
