// uxcore-totop/build/Totop.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _addEventListener = require('rc-util/lib/Dom/addEventListener');

var _addEventListener2 = _interopRequireDefault(_addEventListener);

var _uxcoreAnimate = require('uxcore-animate');

var _uxcoreAnimate2 = _interopRequireDefault(_uxcoreAnimate);

var _TotopBox = require('./TotopBox');

var _TotopBox2 = _interopRequireDefault(_TotopBox);

var _ToTopDefaultBox = require('./ToTopDefaultBox');

var _ToTopDefaultBox2 = _interopRequireDefault(_ToTopDefaultBox);

var _util = require('./util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Totop Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author eternaslky
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2015, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var Totop = function (_React$Component) {
  _inherits(Totop, _React$Component);

  function Totop(props) {
    _classCallCheck(this, Totop);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      showTotop: false
    };
    return _this;
  }

  Totop.prototype.componentDidMount = function componentDidMount() {
    var me = this;
    me.renderComponent();
    me.scrollHandler = (0, _addEventListener2["default"])(window, 'scroll', function () {
      var y = _util2["default"].getWindowScrollY();
      if (y > me.props.distance && !me.state.showTotop) {
        me.setState({
          showTotop: true
        });
      } else if (y <= me.props.distance && me.state.showTotop) {
        me.setState({
          showTotop: false
        });
      }
    });
  };

  Totop.prototype.componentDidUpdate = function componentDidUpdate() {
    this.renderComponent();
  };

  Totop.prototype.componentWillUnmount = function componentWillUnmount() {
    var me = this;
    clearTimeout(me.timer);
    if (me.scrollHandler) {
      me.scrollHandler.remove();
    }
    me.removeContainer();
  };

  Totop.prototype.getContainer = function getContainer() {
    if (!this.container) {
      var getContainer = this.props.getContainer || _util2["default"].defaultGetContainer;
      this.container = getContainer();
    }
    return this.container;
  };

  Totop.prototype.getComponent = function getComponent() {
    var _classnames,
        _this2 = this;

    var me = this;
    var locale = this.props.locale;

    return _react2["default"].createElement(
      'div',
      {
        className: (0, _classnames3["default"])((_classnames = {}, _defineProperty(_classnames, me.props.prefixCls, true), _defineProperty(_classnames, me.props.className, !!me.props.className), _defineProperty(_classnames, me.props.theme, !!me.props.theme), _defineProperty(_classnames, 'fn-clear', true), _classnames))
      },
      _react2["default"].createElement(
        _uxcoreAnimate2["default"],
        { showProp: 'show', transitionName: 'fade' },
        _react2["default"].createElement(_ToTopDefaultBox2["default"], {
          locale: locale,
          show: me.state.showTotop,
          onClick: function onClick() {
            _this2.handleGotopClick();
          }
        })
      ),
      me.props.children
    );
  };

  Totop.prototype.handleGotopClick = function handleGotopClick() {
    var me = this;
    me.scrollTo(me.props.to, me.props.duration, me.props.onTotopEnd);
  };

  Totop.prototype.removeContainer = function removeContainer() {
    if (this.container) {
      var container = this.container;

      _reactDom2["default"].unmountComponentAtNode(container);
      container.parentNode.removeChild(container);
      this.container = null;
    }
  };

  /**
   * scroll method to action like jQuery animation.
   * @param element {DOM} scroll element
   * @param to {number} the final scrollTop you want
   * @param duration {number} scroll animation time (ms)
   */

  Totop.prototype.scrollTo = function scrollTo(to, duration, callback) {
    var me = this;
    if (duration <= 0) {
      if (callback) {
        callback();
      }
      return;
    }
    var y = _util2["default"].getWindowScrollY();
    var difference = to - y;
    var perTick = difference / duration * 10;
    me.timer = setTimeout(function () {
      var targetScrollY = y + perTick;
      _util2["default"].setWindowScrollY(targetScrollY < to ? to : targetScrollY);
      me.scrollTo(to, duration - 10, callback);
    }, 10);
  };

  Totop.prototype.renderComponent = function renderComponent() {
    var me = this;
    var component = me.getComponent();
    _reactDom2["default"].unstable_renderSubtreeIntoContainer(me, component, me.getContainer(), function fallback() {
      me.component = this;
    });
  };

  Totop.prototype.render = function render() {
    return null;
  };

  return Totop;
}(_react2["default"].Component);

Totop.defaultProps = {
  prefixCls: 'kuma-totop',
  to: 10,
  duration: 600,
  distance: 30,
  onTotopEnd: function onTotopEnd() {},
  locale: 'zh-cn'
};

// http://facebook.github.io/react/docs/reusable-components.html
Totop.propTypes = {
  prefixCls: _propTypes2["default"].string,
  className: _propTypes2["default"].string,
  to: _propTypes2["default"].number,
  distance: _propTypes2["default"].number,
  duration: _propTypes2["default"].number,
  onTotopEnd: _propTypes2["default"].func,
  getContainer: _propTypes2["default"].func,
  locale: _propTypes2["default"].string
};

Totop.displayName = 'Totop';

Totop.Box = _TotopBox2["default"];

exports["default"] = Totop;
module.exports = exports['default'];
