// uxcore-steps/build/Steps.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Steps Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author vincent.bian
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2015, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var Steps = function (_React$Component) {
  _inherits(Steps, _React$Component);

  function Steps(props) {
    _classCallCheck(this, Steps);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      tailWidth: 0
    };
    _this.previousStepsWidth = 0;
    _this.itemsWidth = [];
    return _this;
  }

  Steps.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    var _props = this.props,
        type = _props.type,
        direction = _props.direction;

    if (type === 'arrow-bar' || direction === 'vertical') {
      return;
    }

    this.update();

    /*
     * 下面的代码是为了兼容window系统下滚动条出现后会占用宽度的问题。
     * componentDidMount时滚动条还不一定出现了，这时候获取的宽度可能不是最终宽度。
     * 对于滚动条不占用宽度的浏览器，下面的代码也不二次render，resize里面会判断要不要更新。
     */
    setTimeout(function () {
      _this2.resize();
    });

    this.resizeBind = this.resize.bind(this);

    if (window.attachEvent) {
      window.attachEvent('onresize', this.resizeBind);
    } else {
      window.addEventListener('resize', this.resizeBind);
    }
  };

  Steps.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
    if (_react2["default"].Children.count(this.props.children) !== _react2["default"].Children.count(prevProps.children)) {
      this.update();
      return;
    }

    if (this.props.type === 'arrow-bar') {
      return;
    }

    this.resize();
    /*
      * 把最后一个元素设置为absolute，是为了防止动态添加元素后滚动条出现导致的布局问题。
      * 未来不考虑ie8一类的浏览器后，会采用纯css来避免各种问题。
      */
    if (this.props._lastAbsolute) {
      var $dom = this.root;
      var len = $dom.children.length - 1;
      // 可以使用:last-child,但这部分功能应该可以移除,就先不动
      for (var i = 0; i <= len; i++) {
        $dom.children[i].style.position = 'relative';
      }
      $dom.children[len].style.position = 'absolute';
    }
    this.fixLastDetailHeight();
  };

  Steps.prototype.componentWillUnmount = function componentWillUnmount() {
    if (this.props.direction === 'vertical') {
      return;
    }

    if (window.attachEvent) {
      window.detachEvent('onresize', this.resizeBind);
    } else {
      window.removeEventListener('resize', this.resizeBind);
    }
  };

  Steps.prototype.resize = function resize() {
    if (!this.root) {
      return;
    }

    this.fixLastDetailHeight();

    var w = Math.floor(this.root.offsetWidth);
    if (this.previousStepsWidth === w) {
      return;
    }

    this.previousStepsWidth = w;
    this.update();
  };

  /*
   * 把整体高度调整为适合高度,处理最后一个detail是绝对定位的问题
   */


  Steps.prototype.fixLastDetailHeight = function fixLastDetailHeight() {
    if (this.props.type === 'arrow-bar') {
      return;
    }

    var $dom = this.root;
    var len = $dom.children.length - 1;
    var $domLastDetail = $dom.children[len];
    if (this.props.currentDetail === len && $dom.offsetHeight <= $domLastDetail.offsetHeight) {
      $dom.style.height = $domLastDetail.offsetHeight + 'px';
    } else {
      $dom.style.height = 'auto';
    }
  };

  Steps.prototype.update = function update() {
    var $dom = this.root;
    var len = $dom.children.length;
    this.itemsWidth = new Array(len);

    // FIXME: 没太看懂，既然值都一样，为什么还要用一个数组？
    for (var i = 0; i < len; i++) {
      this.itemsWidth[i] = this.props.maxDescriptionWidth;
    }

    this.previousStepsWidth = Math.floor($dom.offsetWidth);

    this.fixLastDetailHeight();

    var tw = this.itemsWidth.reduce(function (prev, w) {
      return prev + w;
    }, 0);
    var dw = Math.floor((this.previousStepsWidth - tw) / (len - 1)) - 1;
    if (dw <= 0) {
      return;
    }

    this.setState({
      tailWidth: dw
    });
  };

  Steps.prototype.render = function render() {
    var _this3 = this;

    var current = this.props.current;
    var _props2 = this.props,
        prefixCls = _props2.prefixCls,
        className = _props2.className,
        children = _props2.children,
        maxDescriptionWidth = _props2.maxDescriptionWidth,
        iconPrefix = _props2.iconPrefix,
        direction = _props2.direction,
        showIcon = _props2.showIcon,
        type = _props2.type,
        showDetail = _props2.showDetail,
        currentDetail = _props2.currentDetail,
        onChange = _props2.onChange;

    var len = _react2["default"].Children.count(children) - 1;
    var iws = this.itemsWidth;
    var clsName = prefixCls;
    var fixStyle = void 0;
    if (direction === 'vertical') {
      clsName += ' ' + prefixCls + '-vertical ' + className;
    } else {
      clsName += ' ' + prefixCls + '-type-' + type + ' ' + className;
      // fix #5
      if (type === 'default') {
        var descItemsCount = children.filter(function (d) {
          return d.props.description;
        }).length;
        if (descItemsCount > 0 && descItemsCount !== len + 1) {
          fixStyle = {
            marginTop: 70
          };
        }
      }
    }
    if (!showIcon) {
      clsName += ' ' + prefixCls + '-noicon';
    }
    if (typeof current !== 'number') {
      current = Number(current);
    }

    return _react2["default"].createElement(
      'div',
      { className: clsName, ref: function ref(c) {
          _this3.root = c;
        } },
      _react2["default"].Children.map(children, function (ele, idx) {
        var np = {
          type: type,
          showIcon: showIcon,
          stepNumber: idx + 1,
          stepLast: idx === len,
          tailWidth: iws.length === 0 || idx === len || direction === 'vertical' ? 'auto' : iws[idx] + _this3.state.tailWidth,
          prefixCls: prefixCls,
          iconPrefix: iconPrefix,
          maxDescriptionWidth: maxDescriptionWidth,
          fixStyle: fixStyle,
          showDetail: showDetail && currentDetail === idx && direction !== 'vertical' && type !== 'long-desc',
          detailContentFixStyle: {
            marginLeft: !isNaN(-(iws[idx] + _this3.state.tailWidth) * idx) ? -(iws[idx] + _this3.state.tailWidth) * idx : 0,
            width: _this3.previousStepsWidth
          },
          onChange: onChange,
          hasDetail: showDetail && direction !== 'vertical' && type !== 'long-desc'
        };
        if (!ele.props.status) {
          if (idx === current) {
            np.status = 'process';
          } else if (idx < current) {
            np.status = 'finish';
          } else {
            np.status = 'wait';
          }
        }

        // If it's vertical, the step point will take 30px width
        if (direction === 'vertical') {
          np.maxDescriptionWidth += 30;
        }

        return _react2["default"].cloneElement(ele, np);
      }, this)
    );
  };

  return Steps;
}(_react2["default"].Component);

Steps.displayName = 'Steps';
Steps.defaultProps = {
  prefixCls: 'kuma-step',
  className: '',
  iconPrefix: '',
  maxDescriptionWidth: 100,
  current: 0,
  direction: '',
  showIcon: true,
  type: 'default',
  showDetail: false,
  currentDetail: 0,
  onChange: function onChange() {},
  children: [],
  // 是否将最后一个元素fix
  _lastAbsolute: false
};
Steps.propTypes = {
  prefixCls: _propTypes2["default"].string,
  className: _propTypes2["default"].string,
  iconPrefix: _propTypes2["default"].string,
  maxDescriptionWidth: _propTypes2["default"].number,
  current: _propTypes2["default"].number,
  direction: _propTypes2["default"].string,
  showIcon: _propTypes2["default"].bool,
  type: _propTypes2["default"].oneOf(['default', 'title-on-top', 'long-desc', 'bottom-desc', 'arrow-bar']),
  showDetail: _propTypes2["default"].bool,
  currentDetail: _propTypes2["default"].number,
  onChange: _propTypes2["default"].func,
  children: _propTypes2["default"].any,
  _lastAbsolute: _propTypes2["default"].bool
};
exports["default"] = Steps;
module.exports = exports['default'];
