// zent/es/tag/Tag.js
import * as React from 'react';
import cx from 'classnames';
import Icon from '../icon';
var PRESET_COLOR = {
    red: true,
    green: true,
    yellow: true,
    blue: true,
    grey: true,
};
export var Tag = React.forwardRef(function (_a, ref) {
    var _b = _a.theme, theme = _b === void 0 ? 'red' : _b, outline = _a.outline, _c = _a.rounded, rounded = _c === void 0 ? true : _c, closable = _a.closable, children = _a.children, className = _a.className, style = _a.style, onClose = _a.onClose, closeButtonStyle = _a.closeButtonStyle, _d = _a.visible, visible = _d === void 0 ? true : _d;
    if (!visible) {
        return null;
    }
    var colorPart = PRESET_COLOR[theme] ? "-" + theme : '';
    var outlinePart = outline ? '-outline' : '';
    return (React.createElement("div", { ref: ref, className: cx('zent-tag', "zent-tag-style" + colorPart + outlinePart, className, {
            'zent-tag-rounded': rounded,
            'zent-tag-closable': closable,
        }), style: style, "data-zv": '7.4.3' },
        React.createElement("div", { className: "zent-tag-content", "data-zv": '7.4.3' }, children),
        closable ? (React.createElement(Icon, { type: "close", className: "zent-tag-close-btn", onClick: onClose, style: closeButtonStyle })) : null));
});
Tag.displayName = 'ZentTag';
export default Tag;
