// formulr/lib/models/field.js
import { __extends } from "tslib";
import { BehaviorSubject } from 'rxjs';
import { BasicModel } from './basic';
import { Some, None, or, isSome, get } from '../maybe';
import { ValidateOption } from '../validate';
import { id } from '../utils';
var FIELD_ID = Symbol('field');
var FieldModel = /** @class */ (function (_super) {
    __extends(FieldModel, _super);
    /** @internal */
    function FieldModel(defaultValue) {
        var _this = _super.call(this) || this;
        _this.defaultValue = defaultValue;
        _this.isTouched = false;
        /**
         * 输入法的 composition 状态
         */
        _this.isCompositing = false;
        /**
         * 用于表单提交前格式化 `Field` 值的回调函数
         */
        _this.normalizeBeforeSubmit = id;
        _this.value$ = new BehaviorSubject(defaultValue);
        return _this;
    }
    Object.defineProperty(FieldModel.prototype, "value", {
        /**
         * 获取 `Field` 当前的值
         */
        get: function () {
            return this.value$.getValue();
        },
        /**
         * 设置 `Field` 的值
         */
        set: function (value) {
            this.value$.next(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 重置 `Field` 为初始值，初始值通过 `initialize` 设置；如果初始值不存在就使用默认值
     */
    FieldModel.prototype.reset = function () {
        this.value$.next(or(this.initialValue, this.defaultValue));
    };
    /**
     * 清除 `Field` 的初始值，并将当前值设置为默认值
     */
    FieldModel.prototype.clear = function () {
        this.initialValue = None();
        this.value$.next(this.defaultValue);
    };
    /**
     * 设置 `Field` 为指定的值，同时会设置初始值
     * @param value 要设置的值
     */
    FieldModel.prototype.initialize = function (value) {
        this.initialValue = Some(value);
        this.value$.next(value);
    };
    FieldModel.prototype.getRawValue = function () {
        return this.value$.getValue();
    };
    /**
     * 获取用于表单提交的值
     */
    FieldModel.prototype.getSubmitValue = function () {
        var normalizeBeforeSubmit = this.normalizeBeforeSubmit;
        return normalizeBeforeSubmit(this.value$.getValue());
    };
    /**
     * `Field` 是否所有校验都通过了
     */
    FieldModel.prototype.valid = function () {
        return this.error$.getValue() == null;
    };
    /**
     * 执行 `Field` 的校验规则
     * @param option 执行校验规则的参数
     */
    FieldModel.prototype.validate = function (option) {
        if (option === void 0) { option = ValidateOption.Default; }
        return this.triggerValidate(option);
    };
    /**
     * 更新 `Field` 的值
     * @param value 要设置的值
     */
    FieldModel.prototype.patchValue = function (value) {
        this.value$.next(value);
    };
    /**
     * `Field` 的值是否没有改变过，如果存在初始值会和初始值比较，否则和默认值比较
     */
    FieldModel.prototype.pristine = function () {
        var value = this.value$.getValue();
        if (isSome(this.initialValue)) {
            return value === get(this.initialValue);
        }
        return value === this.defaultValue;
    };
    /**
     * `Field` 的值是否改变过，如果存在初始值会和初始值比较，否则和默认值比较
     *
     * `dirty === !pristine`
     */
    FieldModel.prototype.dirty = function () {
        return !this.pristine();
    };
    /**
     * 用户是否操作过 `Field`，一般是在 `blur` 事件后设置，部分 `Field` 没有 `blur` 事件可能会在 `change` 的时候设置这个状态
     */
    FieldModel.prototype.touched = function () {
        return this.isTouched;
    };
    return FieldModel;
}(BasicModel));
FieldModel.prototype[FIELD_ID] = true;
function isFieldModel(maybeModel) {
    return !!(maybeModel && maybeModel[FIELD_ID]);
}
export { FieldModel, isFieldModel };
