// formulr/lib/models/form.js
import { __extends } from "tslib";
import { BehaviorSubject } from 'rxjs';
import { FieldSetModel } from './set';
import { ValidateOption } from '../validate';
var FormStrategy;
(function (FormStrategy) {
    /**
     * 指定 model 模式
     */
    FormStrategy[FormStrategy["Model"] = 0] = "Model";
    /**
     * 视图驱动模式
     */
    FormStrategy[FormStrategy["View"] = 1] = "View";
})(FormStrategy || (FormStrategy = {}));
var FORM_ID = Symbol('form');
var FormModel = /** @class */ (function (_super) {
    __extends(FormModel, _super);
    function FormModel(children) {
        var _this = _super.call(this, children) || this;
        _this.children = children;
        /** @internal */
        _this.workingValidators = new Set();
        _this.isValidating$ = new BehaviorSubject(false);
        _this.form = _this;
        var keys = Object.keys(children);
        var keysLength = keys.length;
        for (var index = 0; index < keysLength; index++) {
            var name_1 = keys[index];
            var child = children[name_1];
            _this.registerChild(name_1, child);
        }
        return _this;
    }
    /**
     * 执行整个 `Form` 的校验，会递归触发所有表单元素的校验
     * @param option 表单校验的参数
     */
    FormModel.prototype.validate = function (option) {
        if (option === void 0) { option = ValidateOption.Default; }
        return _super.prototype.validate.call(this, option | ValidateOption.IncludeChildrenRecursively);
    };
    /** @internal */
    FormModel.prototype.addWorkingValidator = function (v) {
        this.workingValidators.add(v);
        this.updateIsValidating();
    };
    /** @internal */
    FormModel.prototype.removeWorkingValidator = function (v) {
        this.workingValidators.delete(v);
        this.updateIsValidating();
    };
    /** @internal */
    FormModel.prototype.updateIsValidating = function () {
        var isValidating = this.workingValidators.size > 0;
        if (isValidating !== this.isValidating$.getValue()) {
            this.isValidating$.next(isValidating);
        }
    };
    return FormModel;
}(FieldSetModel));
FormModel.prototype[FORM_ID] = true;
function isFormModel(maybeModel) {
    return !!(maybeModel && maybeModel[FORM_ID]);
}
export { FormStrategy, FormModel, isFormModel };
