// formulr/lib/models/set.js
import { __extends } from "tslib";
import { Subject } from 'rxjs';
import { BasicModel, isModel } from './basic';
import { ValidateOption } from '../validate';
import { Some, None } from '../maybe';
import { isPlainObject } from '../utils';
var SET_ID = Symbol('set');
var FieldSetModel = /** @class */ (function (_super) {
    __extends(FieldSetModel, _super);
    /** @internal */
    function FieldSetModel(children) {
        var _this = _super.call(this) || this;
        /** @internal */
        _this.patchedValue = null;
        _this.childRegister$ = new Subject();
        _this.childRemove$ = new Subject();
        _this.children = {};
        var keys = Object.keys(children);
        var keysLength = keys.length;
        for (var index = 0; index < keysLength; index++) {
            var name_1 = keys[index];
            var child = children[name_1];
            _this.registerChild(name_1, child);
        }
        _this.children = children;
        return _this;
    }
    /**
     * 初始化 `FieldSet` 的值，并设置 `initialValue`
     * @param values 待初始化的值
     */
    FieldSetModel.prototype.initialize = function (values) {
        if (!isPlainObject(values)) {
            return;
        }
        this.initialValue = Some(values);
        var keys = Object.keys(values);
        for (var i = 0; i < keys.length; i += 1) {
            var key = keys[i];
            var child = this.children[key];
            if (isModel(child)) {
                child.initialize(values[key]);
            }
        }
    };
    /**
     * @internal
     */
    FieldSetModel.prototype.getPatchedValue = function (name) {
        if (this.patchedValue && name in this.patchedValue) {
            return Some(this.patchedValue[name]);
        }
        return None();
    };
    /**
     * 获取 `FieldSet` 的值
     */
    FieldSetModel.prototype.getRawValue = function () {
        var value = {};
        var childrenKeys = Object.keys(this.children);
        for (var i = 0; i < childrenKeys.length; i++) {
            var key = childrenKeys[i];
            var model = this.children[key];
            var childValue = model.getRawValue();
            value[key] = childValue;
        }
        return value;
    };
    /**
     * 获取 `FieldSet` 用于表单提交的值
     */
    FieldSetModel.prototype.getSubmitValue = function () {
        var value = {};
        var childrenKeys = Object.keys(this.children);
        for (var i = 0; i < childrenKeys.length; i++) {
            var key = childrenKeys[i];
            var model = this.children[key];
            var childValue = model.getSubmitValue();
            value[key] = childValue;
        }
        return value;
    };
    /**
     * 在 `FieldSet` 上注册一个新的字段
     * @param name 字段名
     * @param model 字段对应的 model
     */
    FieldSetModel.prototype.registerChild = function (name, model) {
        if (this.children[name]) {
            var prevModel = this.children[name];
            prevModel.form = null;
            prevModel.owner = null;
        }
        model.form = this.form;
        model.owner = this;
        this.children[name] = model;
        this.childRegister$.next(name);
    };
    /**
     * 在 `FieldSet` 上删除指定的字段
     * @param name 字段名
     */
    FieldSetModel.prototype.removeChild = function (name) {
        var model = this.children[name];
        delete this.children[name];
        model.form = null;
        model.owner = null;
        this.childRemove$.next(name);
        return model;
    };
    /**
     * 是否 `FieldSet` 所有字段都通过了校验
     */
    FieldSetModel.prototype.valid = function () {
        if (this.error$.getValue() !== null) {
            return false;
        }
        var keys = Object.keys(this.children);
        for (var i = 0; i < keys.length; i += 1) {
            var key = keys[i];
            var child = this.children[key];
            if (!child.valid()) {
                return false;
            }
        }
        return true;
    };
    /**
     * 更新 `FieldSet` 的值
     * @param value 待更新的值
     */
    FieldSetModel.prototype.patchValue = function (value) {
        if (!isPlainObject(value)) {
            return;
        }
        this.patchedValue = value;
        var keys = Object.keys(value);
        for (var i = 0; i < keys.length; i += 1) {
            var key = keys[i];
            var child = this.children[key];
            if (child) {
                child.patchValue(value[key]);
            }
        }
    };
    /**
     * 清除 `FieldSet` 所有字段的值，同时清除 `initialValue`
     */
    FieldSetModel.prototype.clear = function () {
        var keys = Object.keys(this.children);
        for (var i = 0; i < keys.length; i += 1) {
            var key = keys[i];
            var child = this.children[key];
            if (child) {
                child.clear();
            }
        }
    };
    /**
     * 重置 `FieldValue` 所有字段的值，如果存在 `initialValue` 就是用初始值，否则使用默认值
     */
    FieldSetModel.prototype.reset = function () {
        var keys = Object.keys(this.children);
        for (var i = 0; i < keys.length; i += 1) {
            var key = keys[i];
            var child = this.children[key];
            if (child) {
                child.reset();
            }
        }
    };
    /**
     * 执行 `FieldSet` 的校验
     * @param option 校验的参数
     */
    FieldSetModel.prototype.validate = function (option) {
        var _this = this;
        if (option === void 0) { option = ValidateOption.Default; }
        if (option & ValidateOption.IncludeChildrenRecursively) {
            return Promise.all(Object.keys(this.children)
                .map(function (key) { return _this.children[key].validate(option); })
                .concat(this.triggerValidate(option)));
        }
        return this.triggerValidate(option);
    };
    /**
     * 是否 `FieldSet` 上的所有字段都没有被修改过
     */
    FieldSetModel.prototype.pristine = function () {
        var keys = Object.keys(this.children);
        for (var i = 0; i < keys.length; i += 1) {
            var key = keys[i];
            var child = this.children[key];
            if (!child.pristine()) {
                return false;
            }
        }
        return true;
    };
    /**
     * 是否 `FieldSet` 上有任意字段被修改过
     *
     * `dirty === !pristine`
     */
    FieldSetModel.prototype.dirty = function () {
        return !this.pristine();
    };
    /**
     * 是否 `FieldSet` 上有任意字段被 touch 过
     *
     */
    FieldSetModel.prototype.touched = function () {
        var keys = Object.keys(this.children);
        for (var i = 0; i < keys.length; i += 1) {
            var key = keys[i];
            var child = this.children[key];
            if (child.touched()) {
                return true;
            }
        }
        return false;
    };
    /**
     * 返回指定字段名对应的 model
     * @param name 字段名
     */
    FieldSetModel.prototype.get = function (name) {
        return this.children[name];
    };
    return FieldSetModel;
}(BasicModel));
FieldSetModel.prototype[SET_ID] = true;
function isFieldSetModel(maybeModel) {
    return !!(maybeModel && maybeModel[SET_ID]);
}
export { FieldSetModel, isFieldSetModel };
