// zent/es/popover/placement/auto-bottom-center.js
import getViewportSize from '../../utils/dom/getViewportSize';
import createPlacement from './create';
import BottomLeft from './bottom-left';
import BottomRight from './bottom-right';
import BottomCenter from './bottom-center';
import TopLeft from './top-left';
import TopRight from './top-right';
import TopCenter from './top-center';
var positionMap = {
    BottomLeft: BottomLeft,
    BottomRight: BottomRight,
    BottomCenter: BottomCenter,
    TopLeft: TopLeft,
    TopRight: TopRight,
    TopCenter: TopCenter,
};
var locate = function (anchorBoundingBox, containerBoundingBox, contentDimension, options) {
    var viewport = getViewportSize();
    var anchorBoundingBoxViewport = options.anchorBoundingBoxViewport, cushion = options.cushion;
    var horizontal;
    var vertical;
    var mid = (anchorBoundingBoxViewport.left + anchorBoundingBoxViewport.right) / 2;
    var halfWidth = contentDimension.width / 2;
    if (mid + halfWidth > viewport.width &&
        anchorBoundingBoxViewport.right - contentDimension.width > 0) {
        horizontal = 'Right';
    }
    else if (mid - halfWidth < 0 &&
        anchorBoundingBoxViewport.left + contentDimension.width < viewport.width) {
        horizontal = 'Left';
    }
    else {
        horizontal = 'Center';
    }
    if (anchorBoundingBoxViewport.bottom + cushion + contentDimension.height >
        viewport.height &&
        anchorBoundingBoxViewport.top - cushion - contentDimension.height > 0) {
        vertical = 'Top';
    }
    else {
        vertical = 'Bottom';
    }
    var key = "" + vertical + horizontal;
    return positionMap[key].locate(anchorBoundingBox, containerBoundingBox, contentDimension, options);
};
var AutoBottomCenter = createPlacement(locate);
export default AutoBottomCenter;
