// zent/es/popover/placement/index.js
import create from './create';
import BottomLeft from './bottom-left';
import BottomCenter from './bottom-center';
import BottomRight from './bottom-right';
import LeftTop from './left-top';
import LeftCenter from './left-center';
import LeftBottom from './left-bottom';
import RightTop from './right-top';
import RightCenter from './right-center';
import RightBottom from './right-bottom';
import TopLeft from './top-left';
import TopCenter from './top-center';
import TopRight from './top-right';
import AutoBottomLeft from './auto-bottom-left';
import AutoBottomRight from './auto-bottom-right';
import AutoBottomCenter from './auto-bottom-center';
import AutoTopLeft from './auto-top-left';
import AutoTopRight from './auto-top-right';
import AutoTopCenter from './auto-top-center';
import Arrow from '../arrow-position';
export default {
    create: create,
    Arrow: Arrow,
    BottomLeft: BottomLeft,
    BottomCenter: BottomCenter,
    BottomRight: BottomRight,
    LeftTop: LeftTop,
    LeftCenter: LeftCenter,
    LeftBottom: LeftBottom,
    RightTop: RightTop,
    RightCenter: RightCenter,
    RightBottom: RightBottom,
    TopLeft: TopLeft,
    TopCenter: TopCenter,
    TopRight: TopRight,
    AutoBottomLeft: AutoBottomLeft,
    AutoBottomRight: AutoBottomRight,
    AutoBottomCenter: AutoBottomCenter,
    AutoTopLeft: AutoTopLeft,
    AutoTopRight: AutoTopRight,
    AutoTopCenter: AutoTopCenter,
};
