// zent/es/datetimepicker/year/YearPanelBody.js
import { __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import classNames from 'classnames';
import { CURRENT_YEAR } from '../constants';
import PanelCell from '../common/PanelCell';
var ROW = 4;
var COL = 3;
var YearPanelBody = (function (_super) {
    __extends(YearPanelBody, _super);
    function YearPanelBody() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    YearPanelBody.prototype.getYears = function () {
        var years = [];
        var index = 0;
        var _a = this.props, actived = _a.actived, selected = _a.selected, disabledDate = _a.disabledDate;
        var beginYear = actived.getFullYear() - 4;
        for (var rowIndex = 0; rowIndex < ROW; rowIndex++) {
            years[rowIndex] = [];
            for (var colIndex = 0; colIndex < COL; colIndex++) {
                var val = beginYear + index;
                var isCurrent = val === CURRENT_YEAR;
                var isSelected = selected instanceof Date && selected.getFullYear() === val;
                var isDisabled = disabledDate && disabledDate(val);
                var className = classNames({
                    'year-panel__cell panel__cell': true,
                    'panel__cell--current': isCurrent,
                    'panel__cell--selected': isSelected,
                    'panel__cell--disabled': isDisabled,
                });
                years[rowIndex][colIndex] = {
                    text: val,
                    value: val,
                    title: "" + val,
                    className: className,
                    isDisabled: isDisabled,
                };
                index++;
            }
        }
        return years;
    };
    YearPanelBody.prototype.render = function () {
        var onSelect = this.props.onSelect;
        var years = this.getYears();
        return (React.createElement("div", { className: "year-table panel-table", "data-zv": '7.4.3' },
            React.createElement(PanelCell, { onSelect: onSelect, cells: years })));
    };
    return YearPanelBody;
}(PureComponent));
export default YearPanelBody;
