// zent/es/datetimepicker/month/MonthPanelBody.js
import { __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import classNames from 'classnames';
import PanelCell from '../common/PanelCell';
import { CURRENT_MONTH, CURRENT_YEAR } from '../constants';
var ROW = 4;
var COL = 3;
var MonthPanelBody = (function (_super) {
    __extends(MonthPanelBody, _super);
    function MonthPanelBody() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MonthPanelBody.prototype.isSelected = function (val) {
        var _a = this.props, selected = _a.selected, year = _a.year;
        if (selected && selected instanceof Date) {
            return val === selected.getMonth() && year === selected.getFullYear();
        }
        return false;
    };
    MonthPanelBody.prototype.render = function () {
        var _this = this;
        var _a = this.props, onSelect = _a.onSelect, disabledDate = _a.disabledDate, i18n = _a.i18n, year = _a.year;
        return (React.createElement("div", { className: "month-table panel-table", "data-zv": '7.4.3' },
            React.createElement(PanelCell, { onSelect: onSelect, cells: (function () {
                    var months = [];
                    var index = 0;
                    for (var rowIndex = 0; rowIndex < ROW; rowIndex++) {
                        months[rowIndex] = [];
                        for (var colIndex = 0; colIndex < COL; colIndex++) {
                            var isCurrent = index === CURRENT_MONTH && year === CURRENT_YEAR;
                            var isSelected = _this.isSelected(index);
                            var isDisabled = disabledDate && disabledDate(index);
                            var className = classNames({
                                'panel__cell month-panel__cell': true,
                                'panel__cell--current': isCurrent,
                                'panel__cell--selected': isSelected,
                                'panel__cell--disabled': isDisabled,
                            });
                            months[rowIndex][colIndex] = {
                                text: i18n.panel.monthNames[index],
                                value: index,
                                title: i18n.panel.monthNames[index],
                                className: className,
                                isDisabled: isDisabled,
                            };
                            index++;
                        }
                    }
                    return months;
                })() })));
    };
    return MonthPanelBody;
}(PureComponent));
export default MonthPanelBody;
