// uxcore-form/build/FormRowTitle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _uxcoreTitle = require('uxcore-title');

var _uxcoreTitle2 = _interopRequireDefault(_uxcoreTitle);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; } /**
                                                                                                                                                                                                                   * @author: zhouquan.yezq
                                                                                                                                                                                                                   * @time: 8/7 2015
                                                                                                                                                                                                                   * Form Row Title
                                                                                                                                                                                                                   */


var FormRowTitle = function FormRowTitle(_ref) {
  var _classnames;

  var children = _ref.children,
      className = _ref.className,
      prefixCls = _ref.prefixCls,
      jsxtitle = _ref.jsxtitle,
      type = _ref.type;

  var length = _react2["default"].Children.count(children);
  return _react2["default"].createElement(
    'div',
    {
      className: (0, _classnames3["default"])((_classnames = {}, _defineProperty(_classnames, className, !!className), _defineProperty(_classnames, prefixCls, true), _classnames))
    },
    _react2["default"].createElement(
      _uxcoreTitle2["default"],
      { type: type },
      jsxtitle
    ),
    !!length && _react2["default"].Children.map(children, function (item) {
      return item;
    })
  );
};

FormRowTitle.propTypes = {
  prefixCls: _propTypes2["default"].string,
  jsxtitle: _propTypes2["default"].string,
  className: _propTypes2["default"].string,
  children: _propTypes2["default"].any,
  type: _propTypes2["default"].string
};

FormRowTitle.defaultProps = {
  prefixCls: 'kuma-uxform-row-title',
  jsxtitle: '',
  children: undefined,
  className: '',
  type: 'secondary'
};

FormRowTitle.displayName = 'FormRowTitle';

exports["default"] = FormRowTitle;
module.exports = exports['default'];
