// uxcore-collapse/build/Collapse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _cloneDeep = require('lodash/cloneDeep');

var _cloneDeep2 = _interopRequireDefault(_cloneDeep);

var _reactLifecyclesCompat = require('react-lifecycles-compat');

var _Panel = require('./Panel');

var _Panel2 = _interopRequireDefault(_Panel);

var _util = require('./util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var Collapse = function (_Component) {
  _inherits(Collapse, _Component);

  function Collapse(props) {
    _classCallCheck(this, Collapse);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.getItems = _this.getItems.bind(_this);
    _this.getActivityKey = _this.getActivityKey.bind(_this);
    _this.handleClickItem = _this.handleClickItem.bind(_this);

    var activeKey = props.activeKey,
        accordion = props.accordion;
    var defaultActiveKey = props.defaultActiveKey;
    // If is not accordion mode, then, defaultActiveKey should be an array

    if (!accordion) {
      defaultActiveKey = defaultActiveKey || [];
    }

    _this.state = {
      activeKey: activeKey || defaultActiveKey
    };
    return _this;
  }

  Collapse.getDerivedStateFromProps = function getDerivedStateFromProps(props) {
    if ('activeKey' in props) {
      return {
        activeKey: props.activeKey
      };
    }

    return null;
  };

  Collapse.prototype.getItems = function getItems() {
    var _this2 = this;

    var activeKey = this.getActivityKey();
    var _props = this.props,
        prefixCls = _props.prefixCls,
        accordion = _props.accordion,
        children = _props.children;


    return _react.Children.map(children, function (child, index) {
      // If there is no key provide, use the panel order as default key
      var key = child.key || index;
      var header = child.props.header;

      var isActive = false;
      if (accordion) {
        isActive = activeKey === key;
      } else {
        isActive = activeKey.indexOf(key) > -1;
      }

      var props = {
        key: key,
        header: header,
        isActive: isActive,
        prefixCls: prefixCls,
        children: child.props.children,
        onItemClick: _this2.handleClickItem(key).bind(_this2)
      };

      return _react2["default"].createElement(_Panel2["default"], props);
    });
  };

  Collapse.prototype.getActivityKey = function getActivityKey() {
    var activeKey = this.state.activeKey;
    var accordion = this.props.accordion;

    if (accordion && Array.isArray(activeKey)) {
      activeKey = activeKey[0];
    }

    if (!accordion && !Array.isArray(activeKey)) {
      activeKey = activeKey ? [activeKey] : [];
    }
    return (0, _cloneDeep2["default"])(activeKey);
  };

  Collapse.prototype.handleClickItem = function handleClickItem(key) {
    var _this3 = this;

    return function () {
      var _props2 = _this3.props,
          accordion = _props2.accordion,
          onChange = _props2.onChange;

      var activeKey = _this3.getActivityKey();
      if (accordion) {
        _this3.setState({
          activeKey: key === activeKey ? null : key
        });
      } else {
        var index = activeKey.indexOf(key);
        var isActive = index > -1;
        if (isActive) {
          // remove active state
          activeKey.splice(index, 1);
        } else {
          activeKey.push(key);
        }

        _this3.setState({ activeKey: activeKey });
      }
      onChange(key, activeKey);
    };
  };

  Collapse.prototype.render = function render() {
    var me = this;

    var className = (0, _classnames3["default"])(me.props.prefixCls, _defineProperty({}, me.props.className, !!me.props.className));

    return _react2["default"].createElement(
      'div',
      { className: className },
      me.getItems()
    );
  };

  return Collapse;
}(_react.Component);

Collapse.Panel = _Panel2["default"];
Collapse.displayName = 'Collapse';
Collapse.propTypes = {
  prefixCls: _propTypes2["default"].string,
  className: _propTypes2["default"].string,
  activeKey: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].arrayOf(_propTypes2["default"].string)]),
  defaultActiveKey: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].arrayOf(_propTypes2["default"].string)]),
  onChange: _propTypes2["default"].func,
  accordion: _propTypes2["default"].bool,
  children: _propTypes2["default"].any
};
Collapse.defaultProps = {
  prefixCls: 'kuma-collapse',
  onChange: _util2["default"].noop,
  accordion: false,
  className: '',
  activeKey: undefined,
  defaultActiveKey: undefined,
  children: undefined
};


(0, _reactLifecyclesCompat.polyfill)(Collapse);

exports["default"] = Collapse;
module.exports = exports['default'];
