// uxcore-image/build/Image.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _reactLifecyclesCompat = require('react-lifecycles-compat');

var _addEventListener = require('rc-util/lib/Dom/addEventListener');

var _addEventListener2 = _interopRequireDefault(_addEventListener);

var _utils = require('./utils');

var _cdnAdapter = require('./cdnAdapter');

var _cdnAdapter2 = _interopRequireDefault(_cdnAdapter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Image Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author eternalsky
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2017-2018, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var DEFAULT_TRANS_PIC = "data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='1000' height='1000' style=''%3E %3Crect fill='transparent' width='100%25' height='100%25' x='0' y='0' stroke='none'/%3E %3C/svg%3E";

var Image = function (_React$Component) {
  _inherits(Image, _React$Component);

  function Image(props) {
    _classCallCheck(this, Image);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      renderSrc: (0, _utils.handleImageSrc)(props, _cdnAdapter2["default"]),
      prevSrc: props.src,
      prevEnableUrlAdapter: props.enableUrlAdapter,
      prevAdapterType: props.adapterType,
      loaded: !props.lazyload || (0, _utils.isDocumentLoaded)(),
      showDefault: !props.lazyload
    };
    return _this;
  }

  Image.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    var _props = this.props,
        showDefaultPicDelay = _props.showDefaultPicDelay,
        lazyload = _props.lazyload;

    if (lazyload) {
      if (!(0, _utils.isDocumentLoaded)()) {
        this.loadListener = (0, _addEventListener2["default"])(window, 'load', function () {
          setTimeout(function () {
            _this2.setState({
              loaded: true
            });
          }, 0);
        });
      }

      this.delayTimer = setTimeout(function () {
        _this2.setState({
          showDefault: true
        });
      }, showDefaultPicDelay);
    }
  };

  Image.prototype.componentWillUnmount = function componentWillUnmount() {
    if (this.loadListener) {
      this.loadListener.remove();
    }
    if (this.delayTimer) {
      clearTimeout(this.delayTimer);
    }
  };

  Image.getDerivedStateFromProps = function getDerivedStateFromProps(props, state) {
    if (props.src !== state.prevSrc) {
      return {
        renderSrc: (0, _utils.handleImageSrc)(props, _cdnAdapter2["default"]),
        prevSrc: props.src
      };
    }
    return null;
  };

  Image.prototype.render = function render() {
    var _props2 = this.props,
        className = _props2.className,
        prefixCls = _props2.prefixCls,
        defaultPic = _props2.defaultPic,
        others = _objectWithoutProperties(_props2, ['className', 'prefixCls', 'defaultPic']);

    var _state = this.state,
        renderSrc = _state.renderSrc,
        loaded = _state.loaded,
        showDefault = _state.showDefault;


    var imgSrc = showDefault ? defaultPic : DEFAULT_TRANS_PIC;
    if (loaded) {
      imgSrc = renderSrc;
    }
    var cls = (0, _classnames3["default"])(prefixCls, _defineProperty({}, className, !!className));

    Object.keys(Image.propTypes).forEach(function (key) {
      delete others[key];
    });

    return _react2["default"].createElement('img', _extends({
      alt: ''
    }, others, {
      className: cls,
      src: imgSrc
    }));
  };

  return Image;
}(_react2["default"].Component);

Image.displayName = 'Image';
Image.propTypes = {
  src: _propTypes2["default"].string,
  adapterType: _propTypes2["default"].string,
  className: _propTypes2["default"].string,
  prefixCls: _propTypes2["default"].string,
  defaultPic: _propTypes2["default"].string,
  showDefaultPicDelay: _propTypes2["default"].number,
  enableUrlAdapter: _propTypes2["default"].bool,
  style: _propTypes2["default"].object,
  height: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].number]),
  width: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].number]),
  lazyload: _propTypes2["default"].bool
};
Image.defaultProps = {
  src: '',
  adapterType: '',
  className: '',
  prefixCls: 'uxcore-image',
  defaultPic: "data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='1000' height='1000' style=''%3E %3Crect width='100%25' height='100%25' x='0' y='0' fill='%23f2f3f5' stroke='none'/%3E %3C/svg%3E",
  showDefaultPicDelay: 200,
  enableUrlAdapter: true,
  style: {},
  height: undefined,
  width: undefined,
  lazyload: true
};


Image.adapter = _cdnAdapter2["default"];

(0, _reactLifecyclesCompat.polyfill)(Image);

exports["default"] = Image;
module.exports = exports['default'];
