// zent/es/progress/components/CircleProgress.js
import * as React from 'react';
import { DEFAULT_WIDTH } from '../constants';
import AnimatedArc from './AnimatedArc';
import ProgressInfo from './ProgressInfo';
var CircleProgress = function (props) {
    var percent = props.percent, showInfo = props.showInfo, format = props.format, strokeWidth = props.strokeWidth, width = props.width, bgColor = props.bgColor, color = props.color, state = props.state;
    var progressWidth = width || DEFAULT_WIDTH.CIRCLE;
    var mid = progressWidth / 2;
    var diameter = progressWidth - strokeWidth;
    var radius = diameter / 2;
    var circumference = diameter * Math.PI;
    var offset = (circumference * (100 - percent)) / 100;
    return (React.createElement("div", { className: "zent-progress-container", style: {
            width: progressWidth,
            height: progressWidth,
        }, "data-zv": '7.4.3' },
        React.createElement("div", { className: "zent-progress-wrapper", style: {
                borderWidth: strokeWidth,
                borderColor: bgColor,
                width: progressWidth,
                height: progressWidth,
            }, "data-zv": '7.4.3' }),
        React.createElement("svg", { className: "zent-progress-inner", viewBox: "0 0 " + progressWidth + " " + progressWidth, width: progressWidth, height: progressWidth, "data-zv": '7.4.3' },
            React.createElement("g", { transform: "rotate(-90 " + mid + " " + mid + ")", "data-zv": '7.4.3' },
                React.createElement("circle", { className: "zent-progress-inner-path", cx: mid, cy: mid, r: radius, stroke: color, strokeWidth: strokeWidth, strokeDasharray: circumference, strokeDashoffset: offset, "data-zv": '7.4.3' })),
            state === 'normal' && (React.createElement(AnimatedArc, { className: "zent-progress-path-mask", radius: radius, arcLength: circumference - offset, strokeWidth: strokeWidth }))),
        showInfo && (React.createElement("div", { className: "zent-progress-info", style: {
                lineHeight: progressWidth + "px",
                color: color,
            }, "data-zv": '7.4.3' },
            React.createElement(ProgressInfo, { type: "circle", percent: percent, format: format, state: state })))));
};
export default CircleProgress;
