// uxcore-form/build/FormField/RadioGroupFormField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _uxcoreFormField = require('uxcore-form-field');

var _uxcoreFormField2 = _interopRequireDefault(_uxcoreFormField);

var _uxcoreConst = require('uxcore-const');

var _uxcoreConst2 = _interopRequireDefault(_uxcoreConst);

var _uxcoreRadiogroup = require('uxcore-radiogroup');

var _uxcoreRadiogroup2 = _interopRequireDefault(_uxcoreRadiogroup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Created by xy on 15/4/13.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var Item = _uxcoreRadiogroup2["default"].Item;

var RadioGroupFormField = function (_FormField) {
  _inherits(RadioGroupFormField, _FormField);

  function RadioGroupFormField(props) {
    _classCallCheck(this, RadioGroupFormField);

    var _this = _possibleConstructorReturn(this, _FormField.call(this, props));

    _this.handleChange = _this.handleChange.bind(_this);
    return _this;
  }

  RadioGroupFormField.prototype.handleChange = function handleChange(value) {
    var me = this;
    me.handleDataChange(value);
  };
  /* eslint-disable class-methods-use-this */


  RadioGroupFormField.prototype.addSpecificClass = function addSpecificClass() {
    return 'kuma-radio-group-uxform-field';
  };
  /* eslint-enable class-methods-use-this */

  RadioGroupFormField.prototype.renderField = function renderField() {
    var me = this;
    var arr = [];
    var mode = me.props.jsxmode || me.props.mode;
    var viewModePlaceholder = me.props.viewModePlaceholder || '--';
    if (mode === _uxcoreConst2["default"].MODE.EDIT) {
      arr.push(_react2["default"].createElement(
        _uxcoreRadiogroup2["default"],
        {
          disabled: me.props.jsxdisabled,
          key: 'radiogroup',
          value: me.state.value,
          onChange: me.handleChange
        },
        me.props.children
      ));
    } else {
      var children = me.props.children instanceof Array ? me.props.children : [me.props.children];
      var text = children.filter(function (child) {
        return child.props.value === me.state.value;
      });
      if (text.length) {
        arr.push(_react2["default"].createElement(
          'span',
          { key: 'radiogroup' },
          text[0].props.text || viewModePlaceholder
        ));
      } else {
        arr.push(_react2["default"].createElement(
          'span',
          { key: 'radiogroup' },
          me.state.value || viewModePlaceholder
        ));
      }
    }
    return arr;
  };

  return RadioGroupFormField;
}(_uxcoreFormField2["default"]);

RadioGroupFormField.displayName = 'RadioGroupFormField';
RadioGroupFormField.propTypes = _uxcoreFormField2["default"].propTypes;
RadioGroupFormField.defaultProps = _uxcoreFormField2["default"].defaultProps;
RadioGroupFormField.Item = Item;
exports["default"] = RadioGroupFormField;
module.exports = exports['default'];
