// zent/es/popover/arrow-position/index.js
import create from './create';
import ArrowBottomLeftPosition from './arrow-bottom-left-position';
import ArrowBottomRightPosition from './arrow-bottom-right-position';
import ArrowLeftTopPosition from './arrow-left-top-position';
import ArrowLeftBottomPosition from './arrow-left-bottom-position';
import ArrowRightTopPosition from './arrow-right-top-position';
import ArrowRightBottomPosition from './arrow-right-bottom-position';
import ArrowTopLeftPosition from './arrow-top-left-position';
import ArrowTopRightPosition from './arrow-top-right-position';
export default {
    create: create,
    ArrowBottomLeftPosition: ArrowBottomLeftPosition,
    ArrowBottomRightPosition: ArrowBottomRightPosition,
    ArrowLeftTopPosition: ArrowLeftTopPosition,
    ArrowLeftBottomPosition: ArrowLeftBottomPosition,
    ArrowRightTopPosition: ArrowRightTopPosition,
    ArrowRightBottomPosition: ArrowRightBottomPosition,
    ArrowTopLeftPosition: ArrowTopLeftPosition,
    ArrowTopRightPosition: ArrowTopRightPosition,
};
