// zent/es/menu/MenuItem.js
import { __extends } from "tslib";
import * as React from 'react';
import { Component } from 'react';
import cx from 'classnames';
import { getExtraStyle } from './utils';
var MenuItem = (function (_super) {
    __extends(MenuItem, _super);
    function MenuItem() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleClick = function (e) {
            var _a = _this.props, specKey = _a.specKey, onClick = _a.onClick, disabled = _a.disabled, isInline = _a.isInline, handleSelect = _a.handleSelect;
            if (disabled)
                return;
            if (isInline) {
                handleSelect(specKey);
            }
            onClick(e, specKey);
        };
        return _this;
    }
    MenuItem.prototype.render = function () {
        var _a;
        var _b = this.props, specKey = _b.specKey, prefix = _b.prefix, className = _b.className, children = _b.children, disabled = _b.disabled, isInline = _b.isInline, selectedKey = _b.selectedKey, depth = _b.depth, inlineIndent = _b.inlineIndent;
        var isSelected = selectedKey === specKey;
        var styleObj = getExtraStyle({
            isInline: isInline,
            depth: depth,
            inlineIndent: inlineIndent,
        });
        return (React.createElement("li", { className: cx(prefix + "-menu-item", className, (_a = {},
                _a[prefix + "-menu__inline-item"] = isInline,
                _a[prefix + "-menu__inline-item-selected"] = isSelected,
                _a[prefix + "-menu-item-disabled"] = disabled,
                _a)), style: styleObj, onClick: this.handleClick, "data-zv": '7.4.3' }, children));
    };
    MenuItem.defaultProps = {
        prefix: 'zent',
    };
    return MenuItem;
}(Component));
export { MenuItem };
export default MenuItem;
