// uxcore-table/build/Cell/CheckBox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _uxcoreConst = require('uxcore-const');

var _uxcoreConst2 = _interopRequireDefault(_uxcoreConst);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames3 = require('classnames');

var _classnames4 = _interopRequireDefault(_classnames3);

var _context = require('../context');

var _context2 = _interopRequireDefault(_context);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * A checkbox field
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var CheckBox = function (_React$Component) {
  _inherits(CheckBox, _React$Component);

  function CheckBox(props) {
    _classCallCheck(this, CheckBox);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.handleClick = _this.handleClick.bind(_this);
    return _this;
  }

  CheckBox.prototype.getValue = function getValue() {
    return this.checkbox.checked;
  };

  CheckBox.prototype.handleChange = function handleChange(e) {
    this.props.onChange(e);
  };

  CheckBox.prototype.handleClick = function handleClick() {
    // simulate checkbox
    this.props.onChange({
      target: {
        checked: true
      }
    });
  };

  CheckBox.prototype.renderHalfChecked = function renderHalfChecked() {
    var _this2 = this;

    return _react2["default"].createElement(
      _context2["default"].Consumer,
      null,
      function (context) {
        var _classnames;

        return _react2["default"].createElement(
          'span',
          {
            className: (0, _classnames4["default"])((_classnames = {}, _defineProperty(_classnames, context.prefixCls + '-row-selector', true), _defineProperty(_classnames, _this2.props.className, !!_this2.props.className), _classnames))
          },
          _react2["default"].createElement('i', { className: 'half-checked', onClick: _this2.handleClick })
        );
      }
    );
  };

  CheckBox.prototype.renderCheckBox = function renderCheckBox(renderProps) {
    var _this3 = this;

    return _react2["default"].createElement(
      _context2["default"].Consumer,
      null,
      function (context) {
        var _classnames2;

        return _react2["default"].createElement(
          'label',
          {
            className: (0, _classnames4["default"])((_classnames2 = {}, _defineProperty(_classnames2, context.prefixCls + '-row-selector', true), _defineProperty(_classnames2, _this3.props.className, !!_this3.props.className), _classnames2))
          },
          _react2["default"].createElement('input', _extends({ type: 'checkbox', ref: function ref(c) {
              _this3.checkbox = c;
            } }, renderProps)),
          _react2["default"].createElement('s', null),
          ' ' + renderProps.text || null
        );
      }
    );
  };

  CheckBox.prototype.render = function render() {
    var props = this.props;

    if (props.halfChecked) {
      return this.renderHalfChecked();
    }
    var renderProps = {
      className: 'kuma-checkbox',
      checked: this.props.checked || false,
      onChange: this.handleChange.bind(this),
      text: this.props.text
    };
    if (!!props.disable || props.mode === _uxcoreConst2["default"].MODE.VIEW) {
      renderProps.disabled = true;
    }
    return this.renderCheckBox(renderProps);
  };

  return CheckBox;
}(_react2["default"].Component);

CheckBox.propTypes = {
  onChange: _propTypes2["default"].func,
  checked: _propTypes2["default"].bool,
  className: _propTypes2["default"].string,
  text: _propTypes2["default"].string
};

CheckBox.defaultProps = {
  onChange: function onChange() {},
  checked: undefined,
  className: '',
  text: ''
};

exports["default"] = CheckBox;
module.exports = exports['default'];
