// uxcore-float-nav/build/NavItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames5 = require('classnames');

var _classnames6 = _interopRequireDefault(_classnames5);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _reactLifecyclesCompat = require('react-lifecycles-compat');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * FloatNav Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author vincent.bian
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2015-2016, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var NavItem = function (_Component) {
  _inherits(NavItem, _Component);

  function NavItem(props) {
    _classCallCheck(this, NavItem);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.state = {};
    _this.handleClick = _this.handleClick.bind(_this);
    return _this;
  }

  NavItem.prototype.getSnapshotBeforeUpdate = function getSnapshotBeforeUpdate(prevProps) {
    if (!prevProps.active && this.props.active) {
      this.props.onActive(this.props.anchor, this.root);
    }
    return null;
  };

  NavItem.prototype.componentDidUpdate = function componentDidUpdate(prevProps, prevState, snapshot) {
    // hold for getSnapshotBeforeUpdate method
  };

  NavItem.prototype.handleClick = function handleClick(e) {
    var _props = this.props,
        anchor = _props.anchor,
        onClick = _props.onClick,
        orderNumber = _props.orderNumber,
        onActive = _props.onActive;

    if (!anchor) {
      e.preventDefault();
    }
    onActive(anchor, this.root, 'click');
    onClick(anchor, orderNumber);
  };

  NavItem.prototype.renderSubNav = function renderSubNav() {
    var _props2 = this.props,
        prefixCls = _props2.prefixCls,
        children = _props2.children,
        orderNumber = _props2.orderNumber,
        onActive = _props2.onActive,
        subActiveAnchor = _props2.subActiveAnchor,
        level = _props2.level,
        spotOnly = _props2.spotOnly;

    if (!children || children.length === 0) {
      return null;
    }
    var cloneProps = {
      prefixCls: prefixCls,
      rootNode: false,
      spotOnly: spotOnly
    };
    var itemProps = void 0;

    return _react2["default"].createElement(
      'div',
      { className: (0, _classnames6["default"])(prefixCls + '-subs') },
      _react2["default"].Children.map(children, function (item, index) {
        itemProps = (0, _objectAssign2["default"])({
          onActive: onActive,
          active: item.props.anchor && subActiveAnchor === item.props.anchor,
          subActiveAnchor: subActiveAnchor,
          level: level + 1
        }, cloneProps);
        if (orderNumber) {
          (0, _objectAssign2["default"])(itemProps, {
            orderNumber: orderNumber + '.' + (index + 1)
          });
        }
        return _react2["default"].cloneElement(item, itemProps);
      })
    );
  };

  NavItem.prototype.renderCircleIcon = function renderCircleIcon() {
    var _props3 = this.props,
        rootNode = _props3.rootNode,
        active = _props3.active;

    if (rootNode && !active) {
      return _react2["default"].createElement(
        'svg',
        { width: '20', height: '20' },
        _react2["default"].createElement('circle', { cx: '10', cy: '10', r: '3', className: 'item-circle-icon' })
      );
    }
    return null;
  };

  NavItem.prototype.renderArrowIcon = function renderArrowIcon() {
    /* const { active } = this.props;
    if (active) {
      return (
        <svg width="21" height="20" viewBox="0 0 90 60">
          <polygon points="0,0 60,0 90,30 60,60 0,60" className="item-arrow-icon" />
        </svg>
      );
    }*/
    return null;
  };

  NavItem.prototype.render = function render() {
    var _this2 = this;

    var _props4 = this.props,
        prefixCls = _props4.prefixCls,
        title = _props4.title,
        orderNumber = _props4.orderNumber,
        anchor = _props4.anchor,
        active = _props4.active,
        level = _props4.level,
        spotOnly = _props4.spotOnly;

    if (spotOnly) {
      return _react2["default"].createElement(
        'div',
        {
          className: (0, _classnames6["default"])(prefixCls + '-item', _defineProperty({}, prefixCls + '-item-active', active)), ref: function ref(node) {
            return _this2.root = node;
          }
        },
        _react2["default"].createElement('div', { className: (0, _classnames6["default"])(prefixCls + '-spot-circle', _defineProperty({}, prefixCls + '-spot-circle-active', active)) }),
        this.renderSubNav()
      );
    }
    return _react2["default"].createElement(
      'div',
      {
        className: (0, _classnames6["default"])(prefixCls + '-item', _defineProperty({}, prefixCls + '-item-active', active)), ref: function ref(node) {
          return _this2.root = node;
        }
      },
      this.renderCircleIcon(),
      this.renderArrowIcon(),
      _react2["default"].createElement(
        'a',
        {
          className: (0, _classnames6["default"])(prefixCls + '-item-title', _defineProperty({}, prefixCls + '-item-title-active', active)),
          style: {
            textIndent: level + 1 + 'em'
          },
          href: '#' + anchor,
          onClick: this.handleClick
        },
        _react2["default"].createElement(
          'span',
          { className: prefixCls + '-item-title-inner' },
          orderNumber,
          ' ',
          title
        )
      ),
      this.renderSubNav()
    );
  };

  return NavItem;
}(_react.Component);

NavItem.displayName = 'NavItem';
NavItem.defaultProps = {
  title: '',
  prefixCls: '',
  orderNumber: '',
  anchor: '',
  onClick: function onClick() {},
  onActive: function onActive() {},
  rootNode: true,
  active: false,
  subActiveAnchor: '',
  spotOnly: false
};
NavItem.propTypes = {
  title: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].element]),
  prefixCls: _propTypes2["default"].string,
  orderNumber: _propTypes2["default"].string,
  anchor: _propTypes2["default"].string,
  onClick: _propTypes2["default"].func,
  onActive: _propTypes2["default"].func,
  rootNode: _propTypes2["default"].bool,
  active: _propTypes2["default"].bool,
  subActiveAnchor: _propTypes2["default"].string,
  children: _propTypes2["default"].any,
  level: _propTypes2["default"].number,
  spotOnly: _propTypes2["default"].bool
};
exports["default"] = (0, _reactLifecyclesCompat.polyfill)(NavItem);
module.exports = exports['default'];
