// zent/es/placeholder/shapes/Circle.js
import { __assign, __extends, __rest } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import cx from 'classnames';
import Base from './Base';
var Circle = (function (_super) {
    __extends(Circle, _super);
    function Circle() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Circle.prototype.render = function () {
        var _a = this.props, className = _a.className, prefix = _a.prefix, diameter = _a.diameter, style = _a.style, passThrough = __rest(_a, ["className", "prefix", "diameter", "style"]);
        var classes = cx(prefix + "-placeholder-circle", className);
        var mergedStyle = __assign({ height: diameter, minWidth: diameter }, style);
        return (React.createElement(Base, __assign({ className: classes, prefix: prefix, style: mergedStyle }, passThrough)));
    };
    Circle.defaultProps = {
        style: {},
        prefix: 'zent',
        animate: true,
        diameter: 80,
    };
    return Circle;
}(PureComponent));
export default Circle;
