// zent/es/placeholder/shapes/Rectangle.js
import { __assign, __extends, __rest } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import cx from 'classnames';
import Base from './Base';
var Rectangle = (function (_super) {
    __extends(Rectangle, _super);
    function Rectangle() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rectangle.prototype.render = function () {
        var _a = this.props, className = _a.className, prefix = _a.prefix, width = _a.width, height = _a.height, style = _a.style, passThrough = __rest(_a, ["className", "prefix", "width", "height", "style"]);
        var classes = cx(prefix + "-placeholder-rectangle", className);
        var mergedStyle = __assign({ minWidth: width, height: height }, style);
        return (React.createElement(Base, __assign({ className: classes, prefix: prefix, style: mergedStyle }, passThrough)));
    };
    Rectangle.defaultProps = {
        style: {},
        prefix: 'zent',
        animate: true,
        width: 80,
        height: 80,
    };
    return Rectangle;
}(PureComponent));
export default Rectangle;
