// zent/es/placeholder/shapes/TextRow.js
import { __assign, __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import cx from 'classnames';
var TextRow = (function (_super) {
    __extends(TextRow, _super);
    function TextRow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TextRow.prototype.render = function () {
        var _a;
        var _b = this.props, className = _b.className, lineSpacing = _b.lineSpacing, animate = _b.animate, style = _b.style, prefix = _b.prefix;
        var defaultStyles = {
            marginTop: lineSpacing,
        };
        var classes = cx(prefix + "-placeholder-text-row", prefix + "-placeholder-shape", (_a = {},
            _a[prefix + "-placeholder-shape--animate"] = animate,
            _a), className);
        return React.createElement("div", { className: classes, style: __assign(__assign({}, defaultStyles), style), "data-zv": '7.4.3' });
    };
    TextRow.defaultProps = {
        lineSpacing: '0.7em',
        animate: true,
        prefix: 'zent',
    };
    return TextRow;
}(PureComponent));
export default TextRow;
