// zent/es/placeholder/shapes/TextRowDashed.js
import { __assign, __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import cx from 'classnames';
import sample from '../../utils/sample';
import { DEFAULT_SEGMENTS } from './consts';
var TextRowDashed = (function (_super) {
    __extends(TextRowDashed, _super);
    function TextRowDashed(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            segments: sample(DEFAULT_SEGMENTS),
        };
        return _this;
    }
    TextRowDashed.prototype.render = function () {
        var _a;
        var _b = this.props, className = _b.className, lineSpacing = _b.lineSpacing, animate = _b.animate, segments = _b.segments, style = _b.style, prefix = _b.prefix;
        var defaultStyles = {
            marginTop: lineSpacing,
        };
        var classes = cx(prefix + "-placeholder-text-row-dashed", className);
        var segmengtClasses = cx(prefix + "-placeholder-shape", (_a = {},
            _a[prefix + "-placeholder-shape--animate"] = animate,
            _a));
        var rawSegments = Array.isArray(segments)
            ? segments
            : this.state.segments;
        return (React.createElement("div", { className: classes, style: __assign(__assign({}, defaultStyles), style), "data-zv": '7.4.3' }, rawSegments.map(function (seg, i) { return (React.createElement("div", { key: i, className: prefix + "-placeholder-text-row-dashed-segment", style: { width: seg + "%", paddingLeft: i === 0 ? 0 : '0.3em' }, "data-zv": '7.4.3' },
            React.createElement("div", { className: segmengtClasses, "data-zv": '7.4.3' }))); })));
    };
    TextRowDashed.defaultProps = {
        lineSpacing: '0.7em',
        animate: true,
        prefix: 'zent',
    };
    return TextRowDashed;
}(PureComponent));
export default TextRowDashed;
