// zent/es/datetimepicker/quarter/QuarterPanelBody.js
import { __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import classNames from 'classnames';
import getQuarter from 'date-fns/getQuarter';
import PanelCell from '../common/PanelCell';
var ROW = 2;
var COL = 2;
var QuarterPanelBody = (function (_super) {
    __extends(QuarterPanelBody, _super);
    function QuarterPanelBody() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    QuarterPanelBody.prototype.getQuarters = function () {
        var _a = this.props, disabledDate = _a.disabledDate, selected = _a.selected, actived = _a.actived, i18n = _a.i18n;
        var quarters = [];
        var index = 0;
        for (var rowIndex = 0; rowIndex < ROW; rowIndex++) {
            quarters[rowIndex] = [];
            for (var colIndex = 0; colIndex < COL; colIndex++) {
                var isSelected = selected &&
                    actived.getFullYear() === selected.getFullYear() &&
                    getQuarter(selected) === index + 1;
                var isDisabled = disabledDate && disabledDate(index);
                var className = classNames({
                    'panel__cell quarter-panel__cell': true,
                    'panel__cell--current': false,
                    'panel__cell--selected': isSelected,
                    'panel__cell--disabled': isDisabled,
                });
                quarters[rowIndex][colIndex] = {
                    text: i18n.panel.quarterNames[index],
                    value: index,
                    title: i18n.panel.quarterNames[index],
                    className: className,
                    isDisabled: isDisabled,
                };
                index++;
            }
        }
        return quarters;
    };
    QuarterPanelBody.prototype.render = function () {
        var onSelect = this.props.onSelect;
        var quarters = this.getQuarters();
        return (React.createElement("div", { className: "quarter-table panel-table", "data-zv": '7.4.3' },
            React.createElement(PanelCell, { onSelect: onSelect, cells: quarters })));
    };
    return QuarterPanelBody;
}(PureComponent));
export default QuarterPanelBody;
