// uxcore-calendar/build/fullWeek/WeekTBody.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _calendarFullUtil = require('../calendarFullUtil');

var _LeftTimeLine = require('../LeftTimeLine');

var _LeftTimeLine2 = _interopRequireDefault(_LeftTimeLine);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var DATE_COL_COUNT = 8;

var WeekBody = function (_React$Component) {
  _inherits(WeekBody, _React$Component);

  function WeekBody(props) {
    _classCallCheck(this, WeekBody);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.getTableCell = _this.getTableCell.bind(_this);
    return _this;
  }

  WeekBody.prototype.getTableCell = function getTableCell() {
    var _props = this.props,
        prefixCls = _props.prefixCls,
        weekRender = _props.weekRender,
        slicePiece = _props.slicePiece,
        step = _props.step,
        value = _props.value,
        disabledTime = _props.disabledTime,
        disabledDate = _props.disabledDate,
        onSelect = _props.onSelect,
        onDayHover = _props.onDayHover;


    var newStep = step ? parseInt(step, 10) : 60;

    var now = (0, _moment2["default"])();
    var newValue = value || now;

    var nowDay = newValue.day();
    nowDay = nowDay === 0 ? 7 : nowDay;

    var trDateClass = prefixCls + '-date-tr';
    var tableHtml = [];
    var current = (0, _calendarFullUtil.getTime)(this.props);
    var cloneCurrent = current.clone();

    for (var iIndex = 0; iIndex <= slicePiece + 1; iIndex++) {
      var dateCells = [];
      var computedTime = current.clone();
      var leftTimeLine = '';
      if (iIndex <= slicePiece) {
        leftTimeLine = _react2["default"].createElement(_LeftTimeLine2["default"], { prefixCls: prefixCls, step: newStep, current: computedTime, key: iIndex });
      }

      for (var jIndex = 0; jIndex < DATE_COL_COUNT; jIndex++) {
        if (jIndex === 0 && iIndex <= slicePiece) {
          dateCells.push(leftTimeLine);
        } else {
          var _classnames;

          var gapDay = nowDay - jIndex;
          var dateHtml = null;
          var disableTime = false;
          var disableDate = false;
          current = gapDay > 0 ? (0, _moment2["default"])(computedTime).subtract(gapDay, 'days') : (0, _moment2["default"])(computedTime).add(Math.abs(gapDay), 'days');

          if (disabledDate) {
            if (disabledDate(current, newValue)) {
              disableDate = true;
            }
          }
          if (disabledTime) {
            if (disabledTime(current, newValue)) {
              disableTime = true;
            }
          }

          if (jIndex !== 0 && weekRender) {
            dateHtml = weekRender(current, newValue);
          }

          var disabled = disableTime || disableDate;
          dateCells.push(_react2["default"].createElement(
            'td',
            {
              key: iIndex + '-' + jIndex,
              onClick: disabled ? null : onSelect.bind(null, current),
              onMouseEnter: disabled ? null : onDayHover && onDayHover.bind(null, current),
              role: 'gridcell',
              className: (0, _classnames3["default"])((_classnames = {}, _defineProperty(_classnames, prefixCls + '-date-cell', true), _defineProperty(_classnames, prefixCls + '-prev-date-cell', jIndex < nowDay - 1), _defineProperty(_classnames, prefixCls + '-next-date-cell', jIndex > nowDay - 1), _defineProperty(_classnames, prefixCls + '-current-date-cell', current.isSame(now, 'd')), _defineProperty(_classnames, prefixCls + '-date-disable-cell', disabled), _defineProperty(_classnames, prefixCls + '-date-disable-time', disableTime), _defineProperty(_classnames, prefixCls + '-date-disable-date', disableDate), _classnames))
            },
            dateHtml
          ));
        }
      }
      cloneCurrent = (0, _moment2["default"])(cloneCurrent).add(newStep, 'm');
      current = cloneCurrent;
      tableHtml.push(_react2["default"].createElement(
        'tr',
        { key: iIndex, role: 'row', className: trDateClass },
        dateCells
      ));
    }
    return tableHtml;
  };

  WeekBody.prototype.render = function render() {
    var prefixCls = this.props.prefixCls;

    return _react2["default"].createElement(
      'tbody',
      { className: prefixCls + '-tbody' },
      this.getTableCell()
    );
  };

  return WeekBody;
}(_react2["default"].Component);

exports["default"] = WeekBody;


WeekBody.displayName = 'WeekBody';
WeekBody.defaultProps = {
  prefixCls: '',
  weekRender: null,
  slicePiece: 0,
  step: 60,
  value: {},
  disabledTime: null,
  dataCount: 0,
  onSelect: null,
  onDayHover: null,
  disabledDate: null
};
WeekBody.propTypes = {
  prefixCls: _propTypes2["default"].string,
  weekRender: _propTypes2["default"].func,
  slicePiece: _propTypes2["default"].number,
  step: _propTypes2["default"].number,
  value: _propTypes2["default"].object,
  disabledTime: _propTypes2["default"].func,
  dataCount: _propTypes2["default"].number,
  onSelect: _propTypes2["default"].func,
  onDayHover: _propTypes2["default"].func,
  disabledDate: _propTypes2["default"].func
};
module.exports = exports['default'];
