// uxcore-calendar/build/fullWeek/WeekTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _WeekTHead = require('./WeekTHead');

var _WeekTHead2 = _interopRequireDefault(_WeekTHead);

var _WeekTBody = require('./WeekTBody');

var _WeekTBody2 = _interopRequireDefault(_WeekTBody);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var WeekTable = function (_React$Component) {
  _inherits(WeekTable, _React$Component);

  function WeekTable(props) {
    _classCallCheck(this, WeekTable);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      slicePiece: 0
    };
    _this.getSlicePiece = _this.getSlicePiece.bind(_this);
    return _this;
  }

  WeekTable.prototype.componentDidMount = function componentDidMount() {
    this.getSlicePiece();
  };

  WeekTable.prototype.getSlicePiece = function getSlicePiece() {
    var _props = this.props,
        startHour = _props.startHour,
        step = _props.step,
        endHour = _props.endHour;

    startHour = typeof startHour !== 'undefined' ? parseInt(startHour, 10) : 9;
    endHour = typeof endHour !== 'undefined' ? parseInt(endHour, 10) : 23;
    step = step ? parseInt(step, 10) : 60;
    if (endHour < startHour) {
      endHour = startHour;
    }

    var slicePiece = Math.ceil((endHour - startHour) * 60 / step);
    this.setState({ slicePiece: slicePiece });
  };

  WeekTable.prototype.renderEvents = function renderEvents() {
    var _props2 = this.props,
        scheduleRender = _props2.scheduleRender,
        startHour = _props2.startHour,
        step = _props2.step,
        endHour = _props2.endHour,
        value = _props2.value,
        type = _props2.type;
    var slicePiece = this.state.slicePiece;

    var renderOpts = _extends({
      startHour: startHour,
      step: step,
      endHour: endHour,
      slicePiece: slicePiece,
      type: type,
      current: value
    }, this.props);
    if (scheduleRender) {
      return scheduleRender(renderOpts);
    }
    return null;
  };

  WeekTable.prototype.render = function render() {
    var slicePiece = this.state.slicePiece;
    var prefixCls = this.props.prefixCls;

    var weekCls = prefixCls + '-week';
    return _react2["default"].createElement(
      'div',
      { className: weekCls },
      _react2["default"].createElement(
        'table',
        { className: prefixCls + '-table', cellSpacing: '0', role: 'grid' },
        _react2["default"].createElement(_WeekTHead2["default"], this.props),
        _react2["default"].createElement(_WeekTBody2["default"], _extends({}, this.props, { slicePiece: slicePiece }))
      ),
      this.renderEvents()
    );
  };

  return WeekTable;
}(_react2["default"].Component);

WeekTable.displayName = 'WeekTable';
WeekTable.defaultProps = {
  prefixCls: '',
  weekRender: null,
  startHour: 9,
  endHour: 23,
  step: 60
};
WeekTable.propTypes = {
  prefixCls: _propTypes2["default"].string,
  weekRender: _propTypes2["default"].func,
  startHour: _propTypes2["default"].number,
  endHour: _propTypes2["default"].number,
  step: _propTypes2["default"].number
};
exports["default"] = WeekTable;
module.exports = exports['default'];
