// uxcore-calendar/build/MonthCalendar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _Picker = require('rc-calendar/lib/Picker');

var _Picker2 = _interopRequireDefault(_Picker);

var _MonthCalendar = require('rc-calendar/lib/MonthCalendar');

var _MonthCalendar2 = _interopRequireDefault(_MonthCalendar);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _uxcoreIcon = require('uxcore-icon');

var _uxcoreIcon2 = _interopRequireDefault(_uxcoreIcon);

var _classnames4 = require('classnames');

var _classnames5 = _interopRequireDefault(_classnames4);

var _util = require('./util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var CalendarLocale = {};
var getCalendarContainer = _util2["default"].getCalendarContainer,
    generalizeFormat = _util2["default"].generalizeFormat;


CalendarLocale['zh-cn'] = require('rc-calendar/lib/locale/zh_CN');
CalendarLocale['en-us'] = require('rc-calendar/lib/locale/en_US');

var MonthCalendar = function (_React$Component) {
  _inherits(MonthCalendar, _React$Component);

  function MonthCalendar(props) {
    _classCallCheck(this, MonthCalendar);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {};
    _this.clearValue = _this.clearValue.bind(_this);
    _this.handleChange = _this.handleChange.bind(_this);
    return _this;
  }

  MonthCalendar.prototype.getTriggerNode = function getTriggerNode() {
    return this.trigger;
  };

  MonthCalendar.prototype.getDate = function getDate(date) {
    var me = this;
    var _me$props = me.props,
        timezone = _me$props.timezone,
        locale = _me$props.locale;

    var value = (0, _moment2["default"])(date).locale(locale);
    if (timezone) {
      return value.utcOffset(parseInt(timezone, 10) * 60);
    }
    return value;
  };

  MonthCalendar.prototype.clearValue = function clearValue(e) {
    e.stopPropagation();
    var onSelect = this.props.onSelect;

    onSelect(null, null);
  };

  MonthCalendar.prototype.saveRef = function saveRef(refName) {
    var me = this;
    return function (c) {
      me[refName] = c;
    };
  };

  MonthCalendar.prototype.handleChange = function handleChange(v) {
    var _props = this.props,
        onSelect = _props.onSelect,
        format = _props.format;

    if (v) {
      var date = v.valueOf();
      onSelect(new Date(date), v.format(format));
    } else {
      onSelect(v, v);
    }
  };

  MonthCalendar.prototype.render = function render() {
    var _this2 = this;

    var me = this;
    var p = me.props;
    var calendarOptions = {
      className: (0, _classnames5["default"])(p.className, _defineProperty({}, 'kuma-calendar-' + p.size, !!p.size)),
      style: p.style,
      locale: CalendarLocale[p.locale],
      orient: ['top', 'left'],
      showDateInput: p.showDateInput,
      prefixCls: 'kuma-calendar',
      onChange: p.onChange ? p.onChange : function () {},
      disabledDate: function disabledDate(moment) {
        var range = p.allowedMonthRange;
        if (!range || typeof range === 'string') {
          return false;
        }
        var ret = true;
        var isArray = range.splice && range.length;
        var allowedRanges = isArray ? range : [range];
        for (var i = 0; i < allowedRanges.length; i++) {
          var allowed = allowedRanges[i];
          var start = allowed.start,
              end = allowed.end;

          if (!start && !end) {
            break;
          }
          if (!start && moment.isBefore(end) || !end && moment.isAfter(start)) {
            ret = false;
            break;
          }
          if (moment.isBetween(start, end, 'month', '[]')) {
            ret = false;
            break;
          }
        }
        return ret;
      }
    };
    var pickerOptions = {
      disabled: p.disabled,
      onOpenChange: p.onOpenChange,
      align: p.align,
      transitionName: p.transitionName,
      adjustOrientOnCalendarOverflow: false,
      prefixCls: 'kuma-calendar-picker',
      getCalendarContainer: p.getPopupContainer || getCalendarContainer
    };

    if (p.value) {
      var value = this.getDate(p.value);
      pickerOptions.value = value;
      calendarOptions.defaultValue = value;
    } else {
      pickerOptions.value = null;
      calendarOptions.defaultValue = p.defaultOpenValue ? this.getDate(p.defaultOpenValue) : null;
    }

    if (p.defaultValue) {
      var _value = this.getDate(p.defaultValue);
      calendarOptions.defaultValue = _value;
      pickerOptions.defaultValue = _value;
    } else if (!calendarOptions.defaultValue) {
      calendarOptions.defaultValue = this.getDate(new Date().getTime());
    }

    var calendar = _react2["default"].createElement(_MonthCalendar2["default"], calendarOptions);

    var triggerStyle = {};
    if (p.inputWidth) {
      triggerStyle.width = p.inputWidth;
    }

    var inputClassName = (0, _classnames5["default"])('kuma-input', _defineProperty({}, 'kuma-input-' + p.size + '-size', !!p.size));

    var triggerClassName = (0, _classnames5["default"])('kuma-calendar-picker-input', _defineProperty({}, 'kuma-calendar-picker-input-' + p.size, !!p.size));
    return _react2["default"].createElement(
      _Picker2["default"],
      _extends({
        calendar: calendar,
        onChange: me.handleChange
      }, pickerOptions, {
        dropdownClassName: 'date-picker-dropdown-offset-' + p.size
      }),
      function (_ref) {
        var value = _ref.value;

        var showClear = p.allowClear ? value && !p.disabled : false;
        var newValue = value;
        if (newValue) {
          newValue = newValue.format(generalizeFormat(p.format));
        } else {
          newValue = '';
        }
        return _react2["default"].createElement(
          'span',
          { className: triggerClassName, style: triggerStyle, ref: me.saveRef('trigger') },
          _react2["default"].createElement('input', {
            value: newValue,
            readOnly: true,
            disabled: me.props.disabled,
            placeholder: me.props.placeholder,
            className: inputClassName
          }),
          p.hasTrigger ? _react2["default"].createElement(_uxcoreIcon2["default"], { usei: true, name: 'riqi', className: 'kuma-calendar-trigger-icon ' + (showClear ? 'kuma-calendar-trigger-icon__has-clear' : '') }) : null,
          showClear ? _react2["default"].createElement('i', {
            className: 'uxcore-icon uxicon-biaodanlei-tongyongqingchu kuma-icon-close',
            onClick: _this2.clearValue
          }) : null
        );
      }
    );
  };

  return MonthCalendar;
}(_react2["default"].Component);

MonthCalendar.displayName = 'MonthCalendar';
MonthCalendar.defaultProps = {
  format: 'YYYY-MM',
  placeholder: '请选择月份',
  allowClear: true,
  onSelect: function onSelect() {},

  locale: 'zh-cn',
  transitionName: 'calendarSlideUp',
  align: {
    offset: [0, 0]
  },
  showDateInput: false,
  hasTrigger: true,
  getPopupContainer: undefined,
  inputWidth: undefined
};
MonthCalendar.propTypes = {
  allowClear: _propTypes2["default"].bool,
  format: _propTypes2["default"].string,
  inputWidth: _propTypes2["default"].number,
  placeholder: _propTypes2["default"].string,
  onSelect: _propTypes2["default"].func,
  locale: _propTypes2["default"].string,
  getPopupContainer: _propTypes2["default"].func,
  hasTrigger: _propTypes2["default"].bool,
  showDateInput: _propTypes2["default"].bool,
  align: _propTypes2["default"].object,
  transitionName: _propTypes2["default"].string
};

exports["default"] = MonthCalendar;
module.exports = exports['default'];
