// uxcore-calendar/build/RcYearCalendar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _CalendarMixin = require('rc-calendar/lib/mixin/CalendarMixin');

var _CalendarMixin2 = _interopRequireDefault(_CalendarMixin);

var _CommonMixin = require('rc-calendar/lib/mixin/CommonMixin');

var _CommonMixin2 = _interopRequireDefault(_CommonMixin);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _YearPanel = require('./YearPanel');

var _YearPanel2 = _interopRequireDefault(_YearPanel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var YearCalendar = (0, _createReactClass2["default"])({
  mixins: [_CalendarMixin2["default"], _CommonMixin2["default"]],

  onKeyDown: function onKeyDown(e) {
    // const keyCode = e.keyCode;
    // const ctrlKey = e.ctrlKey || e.metaKey;
    // const stateValue = this.state.value;
    // let value = stateValue;
    // switch (keyCode) {
    //   case KeyCode.DOWN:
    //     value = stateValue.clone();
    //     value.addMonth(3);
    //     break;
    //   case KeyCode.UP:
    //     value = stateValue.clone();
    //     value.addMonth(-3);
    //     break;
    //   case KeyCode.LEFT:
    //     value = stateValue.clone();
    //     if (ctrlKey) {
    //       value.addYear(-1);
    //     } else {
    //       value.addMonth(-1);
    //     }
    //     break;
    //   case KeyCode.RIGHT:
    //     value = stateValue.clone();
    //     if (ctrlKey) {
    //       value.addYear(1);
    //     } else {
    //       value.addMonth(1);
    //     }
    //     break;
    //   case KeyCode.ENTER:
    //     this.onSelect(stateValue);
    //     e.preventDefault();
    //     return 1;
    //   default:
    //     return undefined;
    // }
    // if (value !== stateValue) {
    //   this.setValue(value);
    //   e.preventDefault();
    //   return 1;
    // }
  },
  render: function render() {
    var props = this.props;
    var children = _react2["default"].createElement(_YearPanel2["default"], {
      locale: props.locale,
      disabledDate: props.disabledDate,
      value: this.state.value,
      rootPrefixCls: props.prefixCls,
      onChange: this.setValue,
      onSelect: this.onSelect
    });
    return this.renderRoot({
      children: children,
      className: props.prefixCls + '-year-calendar-container'

    });
  }
});
// import KeyCode from 'rc-util/lib/KeyCode';
exports["default"] = YearCalendar;
module.exports = exports['default'];
