// uxcore-carousel/build/Carousel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactSlick = require('react-slick');

var _reactSlick2 = _interopRequireDefault(_reactSlick);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Carousel Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author eternaslky
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2015, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

// matchMedia polyfill for
// https://github.com/WickyNilliams/enquire.js/issues/82

if (typeof window !== 'undefined') {
  var matchMediaPolyfill = function matchMediaPolyfill() {
    return {
      matches: false,
      addListener: function addListener() {},
      removeListener: function removeListener() {}
    };
  };
  window.matchMedia = window.matchMedia || matchMediaPolyfill;
}

var Carousel = function (_React$Component) {
  _inherits(Carousel, _React$Component);

  function Carousel() {
    _classCallCheck(this, Carousel);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  Carousel.prototype.processChildren = function processChildren() {
    var me = this;
    var length = _react2["default"].Children.count(me.props.children);
    if (length > 0) {
      return _react2["default"].Children.map(me.props.children, function (child) {
        return _react2["default"].createElement(
          'div',
          null,
          child
        );
      });
    }
    return _react2["default"].createElement('div', null);
  };

  Carousel.prototype.render = function render() {
    var _classnames;

    var me = this;

    var _me$props = me.props,
        className = _me$props.className,
        prefixCls = _me$props.prefixCls,
        centerDots = _me$props.centerDots,
        largeArrowsAndDots = _me$props.largeArrowsAndDots,
        props = _objectWithoutProperties(_me$props, ['className', 'prefixCls', 'centerDots', 'largeArrowsAndDots']);

    props.arrows = !!props.arrows;

    return _react2["default"].createElement(
      'div',
      {
        className: (0, _classnames3["default"])((_classnames = {}, _defineProperty(_classnames, '' + prefixCls, true), _defineProperty(_classnames, prefixCls + '-dots-centered', !!centerDots), _defineProperty(_classnames, prefixCls + '-large-mode', !!largeArrowsAndDots), _defineProperty(_classnames, prefixCls + '-always-show-arrows', me.props.arrows === 'always'), _defineProperty(_classnames, className, !!className), _classnames))
      },
      _react2["default"].createElement(
        _reactSlick2["default"],
        props,
        me.processChildren()
      )
    );
  };

  return Carousel;
}(_react2["default"].Component);

Carousel.displayName = 'Carousel';
Carousel.propTypes = {
  centerDots: _propTypes2["default"].bool,
  children: _propTypes2["default"].node
};
Carousel.defaultProps = {
  prefixCls: 'kuma-carousel',
  largeArrowsAndDots: false,
  dots: true,
  cssEase: 'ease-out',
  infinite: true,
  speed: 500,
  slidesToShow: 1,
  centerDots: true,
  slidesToScroll: 1,
  arrows: 'hover'
};
exports["default"] = Carousel;
module.exports = exports['default'];
