// uxcore-select-form-field/build/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

/**
 * transfer { a: 'A' } to [{value: 'a', text: 'A'}]
 * transfer [{key: 'a', label: 'A'}] to [{value: 'a', text: 'A'}]
 */
var processData = function processData(data) {
  var values = [];
  if ((typeof data === 'undefined' ? 'undefined' : _typeof(data)) === 'object' && !(data instanceof Array)) {
    var keys = Object.keys(data);
    values = keys.map(function (key) {
      return {
        value: key,
        text: data[key],
        title: data.title || data.label || data.text
      };
    });
  } else {
    values = data.map(function (item) {
      var newItem = _extends({}, item, {
        value: item.key || item.value,
        text: item.label || item.text,
        title: item.title || item.label || item.text
      });
      ['key', 'label'].forEach(function (key) {
        delete newItem[key];
      });
      return newItem;
    });
  }
  return values;
};

/**
 * transfer [{value: 'a', text: 'A'}] to { a: 'A' }
 */
var transferDataToObj = function transferDataToObj(data) {
  var obj = {};
  data.forEach(function (item) {
    var key = item.value === '' ? '__all__' : item.value;
    obj[key] = item.text;
  });
  return obj;
};

/**
 * valueProp maybe props.value or key
 */
var getValuePropValue = function getValuePropValue(child) {
  var key = '';
  if ('value' in child.props) {
    key = child.props.value;
  } else {
    key = child.key;
  }
  return key;
};

exports["default"] = {
  processData: processData,
  transferDataToObj: transferDataToObj,
  getValuePropValue: getValuePropValue
};
module.exports = exports['default'];
