// uxcore-uploader/build/Progress.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uxcoreProgress = require('uxcore-progress');

var _uxcoreIcon = require('uxcore-icon');

var _uxcoreIcon2 = _interopRequireDefault(_uxcoreIcon);

var _uploadcore = require('uploadcore');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var Progress = function (_React$Component) {
  _inherits(Progress, _React$Component);

  function Progress(props) {
    _classCallCheck(this, Progress);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      percentage: 0
    };
    return _this;
  }

  Progress.prototype.onPending = function onPending() {
    this.props.file.pending();
  };

  Progress.prototype.componentDidMount = function componentDidMount() {
    var t = null;
    var me = this;
    var percentage = me.state.percentage;
    me._isMounted = true;
    this.props.file.on(_uploadcore.Events.FILE_UPLOAD_PROGRESS, function (progress) {
      if (me._isMounted) {
        me.setState({
          percentage: progress.percentage
        });
      }
    });
    // me.t = setInterval(() => {
    //     console.log(me.props.file.progress);
    //     percentage = percentage + 5;
    //     if (me.props.isVisual && me.props.status === 'error') {
    //         clearInterval(me.t);
    //     } else {
    //         if (me._isMounted) {
    //             me.setState({
    //                 percentage: percentage
    //             });
    //         }
    //         if (percentage === 95) {
    //             clearInterval(me.t);
    //         }
    //     }
    //
    // }, me.props.interval);
  };

  Progress.prototype.componentWillUnmount = function componentWillUnmount() {
    var me = this;
    me._isMounted = false;
    clearInterval(me.t);
  };

  Progress.prototype.render = function render() {
    if (this.props.isVisual) {
      if (this.props.status === 'error') {
        return _react2["default"].createElement(
          'div',
          { className: 'visual-progress-box' },
          _react2["default"].createElement(_uxcoreIcon2["default"], { onClick: this.props.onPending.bind(this), className: 're-upload-icon', name: 'shuaxin' }),
          _react2["default"].createElement(
            'span',
            { onClick: this.props.onPending.bind(this), className: 're-upload' },
            '\u91CD\u65B0\u4E0A\u4F20'
          ),
          _react2["default"].createElement(
            'div',
            { className: 'delete-progress', onClick: this.props.onCancel.bind(this) },
            _react2["default"].createElement(_uxcoreIcon2["default"], { name: 'biaodanlei-tongyongqingchu' })
          )
        );
      }
      return _react2["default"].createElement(
        'div',
        { className: 'visual-progress-box' },
        _react2["default"].createElement(_uxcoreProgress.Line, { percent: this.state.percentage, strokeWidth: 4, showInfo: false }),
        _react2["default"].createElement(
          'div',
          { className: 'delete-progress', onClick: this.props.onCancel.bind(this) },
          _react2["default"].createElement(_uxcoreIcon2["default"], { name: 'biaodanlei-tongyongqingchu' })
        )
      );
    }
    return _react2["default"].createElement('div', {
      style: {
        width: '100%',
        transform: 'scale(' + this.state.percentage / 100 + ', 1)',
        transformOrigin: 'left top',
        transition: 'transform 0.1s linear'
      }, className: 'progress-box'
    });
  };

  return Progress;
}(_react2["default"].Component);

Progress.propTypes = {
  percentage: _propTypes2["default"].number,
  interval: _propTypes2["default"].number
};
Progress.defaultProps = {
  percentage: 0,
  interval: 100
};
exports["default"] = Progress;
module.exports = exports['default'];
