// uxcore-uploader/build/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _uploadcore = require('uploadcore');

var hijackEvents = ['fileuploadsuccess', 'filecancel'];
var CORE_INSTANCE = {};
_uploadcore.UploadCore.setSWF('https://alinw.alicdn.com/alinw/uxuploader/2.0.1/flashpicker.swf');
exports["default"] = {
  humanSizeFormat: function humanSizeFormat(size) {
    size = parseFloat(size);
    var prefixesSI = ['', 'k', 'm', 'g', 't', 'p', 'e', 'z', 'y'];
    var base = 1000,
        index = size ? Math.floor(Math.log(size) / Math.log(base)) : 0;
    index = Math.min(index, prefixesSI.length - 1);
    var powedPrecision = Math.pow(10, index < 2 ? 0 : index > 2 ? 2 : 1);
    size /= Math.pow(base, index);
    size = Math.round(size * powedPrecision) / powedPrecision;
    if (size > 500) {
      size = Math.round(size / 100) / 10;
      index++;
    }
    return size + prefixesSI[index];
  },
  natcut: function natcut(title, len) {
    var max = len * 2,
        length = title.length,
        l = 0,
        i = 0,
        part = void 0,
        s = void 0;
    for (i = 0; i < length && l <= max; i++) {
      l += title.charCodeAt(i) > 255 ? 2 : 1;
    }
    if (l <= max) {
      return title;
    }
    i = 0;
    l = 0;
    len -= 2;
    while (l < len) {
      s = title.charCodeAt(i) > 255 ? 2 : 1;
      if (l + s > len) {
        break;
      } else {
        i++;
        l += s;
      }
    }
    part = title.substr(0, i);
    l += 3;

    i = length;
    while (l < max) {
      s = title.charCodeAt(i - 1) > 255 ? 2 : 1;
      if (l + s > max) {
        break;
      } else {
        i--;
        l += s;
      }
    }
    return part + '...' + title.substring(Math.min(i, length - 1), length);
  },

  getCoreInstance: function getCoreInstance(props, autoPending) {
    var core = props.core;
    if (core instanceof _uploadcore.UploadCore) {
      return core;
    }

    var id = core;
    if (id && typeof id === 'string' && Object.prototype.hasOwnProperty.call(CORE_INSTANCE, id)) {
      return CORE_INSTANCE[id];
    }

    var options = props.options || {};
    ['name', 'url', 'params', 'action', 'data', 'headers', 'withCredentials', 'timeout', 'chunkEnable', 'chunkSize', 'chunkRetries', 'chunkProcessThreads', 'processThreads', 'autoPending', 'multiple', 'accept', 'sizeLimit', 'preventDuplicate', 'isOnlyImg'].forEach(function (key) {
      if (Object.prototype.hasOwnProperty.call(props, key)) {
        options[key] = props[key];
      }
    });

    if (autoPending != null) {
      options.autoPending = autoPending;
    }

    core = new _uploadcore.UploadCore(options);

    for (var key in props) {
      if (Object.prototype.hasOwnProperty.call(props, key)) {
        var m = /^on(\w+)$/i.exec(key);
        if (!m) continue;
        if (typeof props[key] === 'function' && hijackEvents.indexOf(m[1]) === -1) {
          core.on(m[1], props[key]);
        }
      }
    }

    if (id) {
      CORE_INSTANCE[id] = core;
    }

    return core;
  },
  /**
   * JSON can not handle undefined/null/function
   * hence it's just a simple method
   * Don't use if your object contains the type metioned before
   */
  simpleDeepCopy: function simpleDeepCopy(obj) {
    return JSON.parse(JSON.stringify(obj));
  },
  simpleDeepEqual: function simpleDeepEqual(a, b) {
    return JSON.stringify(a) === JSON.stringify(b);
  },
  TRANSFORM_PROPERTY: function TRANSFORM_PROPERTY() {
    var style = document.createElement('div').style;
    var properties = ['transform', 'WebkitTransform', 'MozTransform', 'msTransform'];
    for (var i = 0, l = properties.length; i < l; i++) {
      if (properties[i] in style) {
        return properties[i];
      }
    }
    return false;
  },

  getUrl: function getUrl(json) {
    try {
      var data = json;
      if (json.content) {
        if (json.content.data) {
          data = json.content.data;
        } else {
          data = json.content;
        }
      } else if (json.data) {
        data = json.data;
      }
      var downloadUrl = data.downloadUrl || data.file || data.url;
      var previewUrl = data.previewUrl || downloadUrl;
      return { downloadUrl: downloadUrl, previewUrl: previewUrl };
    } catch (e) {
      console.error('data or content not found in response, maybe you should set response by yourself');
      return null;
    }
  }
};
module.exports = exports['default'];
