// zent/es/pagination/components/button/ArrowButton.js
import { __assign, __extends, __rest } from "tslib";
import * as React from 'react';
import { Component } from 'react';
import cx from 'classnames';
import Button from '../../../button';
import Pop from '../../../pop';
var XML_NS = 'http://www.w3.org/2000/svg';
var DEFAULT_DISABLED_POP_PROPS = {
    position: 'top-right',
    trigger: 'hover',
    centerArrow: true,
};
var ArrowButton = (function (_super) {
    __extends(ArrowButton, _super);
    function ArrowButton() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ArrowButton.prototype.render = function () {
        var _a = this.props, direction = _a.direction, double = _a.double, active = _a.active, bordered = _a.bordered, disabledHelp = _a.disabledHelp, rest = __rest(_a, ["direction", "double", "active", "bordered", "disabledHelp"]);
        if (double) {
            return (React.createElement(DoubleArrowButton, __assign({ direction: direction, active: active, bordered: bordered }, rest)));
        }
        var Arrow = null;
        if (direction === 'left') {
            Arrow = LeftArrow;
        }
        else if (direction === 'right') {
            Arrow = RightArrow;
        }
        var btn = (React.createElement(Button, __assign({}, rest, { className: cx('zent-pagination-arrow-button', {
                'zent-pagination-button--layout': !disabledHelp,
                'zent-pagination-page-button--no-border': !bordered,
            }) }),
            React.createElement(Arrow, null)));
        if (disabledHelp && rest.disabled) {
            var wrapperClassName = disabledHelp.wrapperClassName, otherProps = __rest(disabledHelp, ["wrapperClassName"]);
            var popProps = __assign(__assign({ wrapperClassName: cx('zent-pagination-button--layout', wrapperClassName) }, DEFAULT_DISABLED_POP_PROPS), otherProps);
            return React.createElement(Pop, __assign({}, popProps), btn);
        }
        return btn;
    };
    ArrowButton.defaultProps = {
        double: false,
        active: false,
        bordered: true,
    };
    return ArrowButton;
}(Component));
export { ArrowButton };
var DoubleArrowButton = (function (_super) {
    __extends(DoubleArrowButton, _super);
    function DoubleArrowButton() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            showArrow: false,
            prevActive: false,
        };
        _this.onMouseOver = function () {
            _this.setState({
                showArrow: true,
            });
        };
        _this.onMouseOut = function () {
            _this.setState({
                showArrow: false,
            });
        };
        return _this;
    }
    DoubleArrowButton.prototype.render = function () {
        var _a = this.props, direction = _a.direction, active = _a.active, bordered = _a.bordered, _ = _a.double, rest = __rest(_a, ["direction", "active", "bordered", "double"]);
        var showArrow = this.state.showArrow;
        var Arrow = null;
        if (direction === 'left') {
            Arrow = LeftDoubleArrow;
        }
        else if (direction === 'right') {
            Arrow = RightDoubleArrow;
        }
        var isActive = showArrow;
        return (React.createElement(Button, __assign({ className: cx('zent-pagination-arrow-button', 'zent-pagination-arrow-button--double', {
                'zent-pagination-arrow-button--double-active': isActive,
                'zent-pagination-page-button--no-border': !bordered,
            }) }, rest, { onMouseOver: active ? undefined : this.onMouseOver, onMouseLeave: active ? undefined : this.onMouseOut }), isActive ? React.createElement(Arrow, null) : '...'));
    };
    DoubleArrowButton.getDerivedStateFromProps = function (props, state) {
        var active = props.active;
        var stateDiff = {
            prevActive: active,
            showArrow: state.showArrow,
        };
        if (active !== state.prevActive) {
            stateDiff.showArrow = !!active;
        }
        return stateDiff;
    };
    DoubleArrowButton.defaultProps = {
        double: false,
        active: false,
        bordered: true,
    };
    return DoubleArrowButton;
}(Component));
function LeftArrow() {
    return (React.createElement("svg", { width: "9", height: "14", xmlns: XML_NS, className: "zent-pagination-arrow", "data-zv": '7.4.3' },
        React.createElement("path", { d: "M8 1L1.78 7 8 13", strokeWidth: 2, fill: "none", fillRule: "evenodd", "data-zv": '7.4.3' })));
}
function RightArrow() {
    return (React.createElement("svg", { width: "9", height: "14", xmlns: XML_NS, className: "zent-pagination-arrow", "data-zv": '7.4.3' },
        React.createElement("path", { d: "M1 1l6.22 6L1 13", strokeWidth: 2, fill: "none", fillRule: "evenodd", "data-zv": '7.4.3' })));
}
function LeftDoubleArrow() {
    return (React.createElement("svg", { width: "15", height: "14", xmlns: XML_NS, className: "zent-pagination-arrow", "data-zv": '7.4.3' },
        React.createElement("g", { strokeWidth: 2, fill: "none", fillRule: "evenodd", "data-zv": '7.4.3' },
            React.createElement("path", { d: "M8 1L1.78 7 8 13", "data-zv": '7.4.3' }),
            React.createElement("path", { d: "M14 1L7.78 7 14 13", "data-zv": '7.4.3' }))));
}
function RightDoubleArrow() {
    return (React.createElement("svg", { width: "15", height: "14", xmlns: XML_NS, className: "zent-pagination-arrow", "data-zv": '7.4.3' },
        React.createElement("g", { strokeWidth: 2, fill: "none", fillRule: "evenodd", "data-zv": '7.4.3' },
            React.createElement("path", { d: "M7 1l6.22 6L7 13", "data-zv": '7.4.3' }),
            React.createElement("path", { d: "M1 1l6.22 6L1 13", "data-zv": '7.4.3' }))));
}
export default ArrowButton;
