// uxcore-cell-field/build/createCellField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _CellField2 = require('./CellField');

var _CellField3 = _interopRequireDefault(_CellField2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var Input = function Input(props) {
  return _react2["default"].createElement('input', _extends({ className: 'kuma-input' }, props, { value: props.value, onChange: function onChange(e) {
      props.onChange(e.target.value);
    } }));
};

Input.propTypes = {
  value: _propTypes2["default"].string,
  onChange: _propTypes2["default"].func
};

var createCellField = function createCellField(options) {
  var CustomField = function (_CellField) {
    _inherits(CustomField, _CellField);

    function CustomField() {
      _classCallCheck(this, CustomField);

      return _possibleConstructorReturn(this, _CellField.apply(this, arguments));
    }

    CustomField.prototype.renderContent = function renderContent() {
      var _extends2;

      var me = this;
      var defaultOptions = {
        valuePropName: 'value',
        changePropName: 'onChange',
        component: _react2["default"].createElement(Input, null),
        processValue: function processValue(value) {
          return value;
        },
        processText: function processText(value) {
          return JSON.stringify(value);
        }
      };
      var newOptions = (0, _objectAssign2["default"])({}, defaultOptions, options);
      return _react2["default"].cloneElement(newOptions.component, _extends((_extends2 = {}, _defineProperty(_extends2, newOptions.valuePropName, me.props.value), _defineProperty(_extends2, newOptions.changePropName, function () {
        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        me.handleDataChange({
          value: newOptions.processValue.apply(me, args),
          text: newOptions.processText.apply(me, args)
        });
      }), _extends2), me.props.column.config || {}));
    };

    return CustomField;
  }(_CellField3["default"]);

  return CustomField;
};

exports["default"] = createCellField;
module.exports = exports['default'];
