// zent/es/datetimepicker/time/HourPanel.js
import { __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import classNames from 'classnames';
import PanelHeader from '../common/PanelHeader';
import TimeCell from './TimeCell';
import { padLeft } from '../utils';
import { CURRENT } from '../constants';
var ROW = 4;
var COL = 7;
var HourPanel = (function (_super) {
    __extends(HourPanel, _super);
    function HourPanel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HourPanel.prototype.isSelected = function (val) {
        var selected = this.props.selected;
        return selected.getHours() === val;
    };
    HourPanel.prototype.isCurrent = function (val) {
        return CURRENT.getHours() === val;
    };
    HourPanel.prototype.getHours = function () {
        var cells = [];
        var i = 0;
        for (var j = 0; j < ROW; j++) {
            for (var k = 0; k < COL && i < 24; k++) {
                var isDisabled = this.props.isDisabled && this.props.isDisabled(i);
                var isSelected = this.isSelected(i);
                var isCurrent = this.isCurrent(i);
                var className = classNames({
                    'panel__cell time-panel__cell': true,
                    'panel__cell--disabled': isDisabled,
                    'panel__cell--selected': isSelected,
                    'panel__cell--current': isCurrent,
                });
                cells[j] = cells[j] || [];
                cells[j][k] = {
                    text: padLeft(i),
                    value: i,
                    isDisabled: isDisabled,
                    className: className,
                };
                i += this.props.step || 1;
            }
        }
        return cells;
    };
    HourPanel.prototype.render = function () {
        var _a = this.props, hidePanel = _a.hidePanel, i18n = _a.i18n, onSelect = _a.onSelect, className = _a.className, hideHeader = _a.hideHeader;
        var hours = this.getHours();
        return (React.createElement("div", { className: classNames('hour-panel', className), "data-zv": '7.4.3' },
            !hideHeader && (React.createElement(PanelHeader, { title: i18n.panel.hourSelect, showNext: false, prev: hidePanel })),
            React.createElement("div", { className: "hour-table panel-table", "data-zv": '7.4.3' },
                React.createElement(TimeCell, { cells: hours, onSelect: onSelect }))));
    };
    return HourPanel;
}(PureComponent));
export default HourPanel;
