// zent/es/datetimepicker/time/SecondPanel.js
import { __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import classNames from 'classnames';
import PanelHeader from '../common/PanelHeader';
import TimeCell from './TimeCell';
import { padLeft } from '../utils';
import { CURRENT } from '../constants';
var ROW = 9;
var COL = 7;
var SecondPanel = (function (_super) {
    __extends(SecondPanel, _super);
    function SecondPanel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SecondPanel.prototype.isSelected = function (val) {
        var selected = this.props.selected;
        return selected.getSeconds() === val;
    };
    SecondPanel.prototype.isCurrent = function (val) {
        return CURRENT.getSeconds() === val;
    };
    SecondPanel.prototype.getSeconds = function () {
        var cells = [];
        var i = 0;
        for (var j = 0; j < ROW; j++) {
            for (var k = 0; k < COL && i < 60; k++) {
                var isDisabled = this.props.isDisabled && this.props.isDisabled(i);
                var isSelected = this.isSelected(i);
                var isCurrent = this.isCurrent(i);
                var className = classNames({
                    'panel__cell time-panel__cell': true,
                    'panel__cell--disabled': isDisabled,
                    'panel__cell--selected': isSelected,
                    'panel__cell--current': isCurrent,
                });
                cells[j] = cells[j] || [];
                cells[j][k] = {
                    text: padLeft(i),
                    value: i,
                    isDisabled: isDisabled,
                    className: className,
                };
                i += this.props.step || 1;
            }
        }
        return cells;
    };
    SecondPanel.prototype.render = function () {
        var _a = this.props, hidePanel = _a.hidePanel, onSelect = _a.onSelect, i18n = _a.i18n, className = _a.className, hideHeader = _a.hideHeader;
        var seconds = this.getSeconds();
        return (React.createElement("div", { className: classNames('second-panel', className), "data-zv": '7.4.3' },
            !hideHeader && (React.createElement(PanelHeader, { title: i18n.panel.secondSelect, showNext: false, prev: hidePanel })),
            React.createElement("div", { className: "second-table panel-table", "data-zv": '7.4.3' },
                React.createElement(TimeCell, { cells: seconds, onSelect: onSelect }))));
    };
    return SecondPanel;
}(PureComponent));
export default SecondPanel;
