// zent/es/datetimepicker/constants/index.js
import Popover from '../../popover';
export function noop() {
    var _ = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        _[_i] = arguments[_i];
    }
}
export var CURRENT = new Date();
export var CURRENT_DAY = new Date(CURRENT.getFullYear(), CURRENT.getMonth(), CURRENT.getDate());
export var CURRENT_YEAR = CURRENT.getFullYear();
export var CURRENT_MONTH = CURRENT.getMonth();
export var CURRENT_DATE = CURRENT.getDate();
export var ONEDAY = 24 * 60 * 60 * 1000;
export var TIME_BEGIN = '00:00:00';
export var TIME_END = '23:59:59';
export var TIME_FORMAT = 'HH:mm:ss';
export var timeFnMap = {
    hour: 'setHours',
    minute: 'setMinutes',
    second: 'setSeconds',
};
export var popPositionMap = {
    left: Popover.Position.AutoBottomLeft,
    right: Popover.Position.AutoBottomRight,
};
export var commonProps = {
    prefix: 'zent',
    confirmText: '',
    format: 'YYYY-MM-DD',
    popPosition: 'left',
    openPanel: false,
    onChange: noop,
    isFooterVisible: false,
    canClear: true,
    autoComplete: 'off',
};
