// uxcore-calendar/build/date/DateTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _DateTHead = require('./DateTHead');

var _DateTHead2 = _interopRequireDefault(_DateTHead);

var _DateTBody = require('./DateTBody');

var _DateTBody2 = _interopRequireDefault(_DateTBody);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } // customized rc-calendar https://github.com/react-component/calendar/blob/master/

var DateTable = function (_React$Component) {
  _inherits(DateTable, _React$Component);

  function DateTable(props) {
    _classCallCheck(this, DateTable);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      tableHeight: 0
    };
    return _this;
  }

  DateTable.prototype.componentDidMount = function componentDidMount() {
    this.setState({
      tableHeight: this.fullTable.offsetHeight
    });
  };

  DateTable.prototype.renderEvents = function renderEvents() {
    var _props = this.props,
        scheduleRender = _props.scheduleRender,
        startHour = _props.startHour,
        step = _props.step,
        endHour = _props.endHour,
        value = _props.value,
        type = _props.type;
    var tableHeight = this.state.tableHeight;


    var renderOpts = _extends({
      startHour: startHour, step: step, endHour: endHour, type: type, current: value }, this.props);
    if (scheduleRender) {
      return scheduleRender(renderOpts, tableHeight);
    }
    return '';
  };

  DateTable.prototype.render = function render() {
    var _classnames,
        _this2 = this;

    var prefixCls = this.props.prefixCls;

    var fullTableHeight = this.fullTable ? this.fullTable.offsetHeight : 0;
    var cellHeight = 0.8 * (fullTableHeight - 32) / 6;
    var tableCls = (0, _classnames3["default"])((_classnames = {}, _defineProperty(_classnames, prefixCls + '-table', true), _defineProperty(_classnames, 'super-mini', cellHeight <= 42 && cellHeight > 0), _classnames));
    return _react2["default"].createElement(
      'div',
      { className: tableCls, key: prefixCls + '-date' },
      _react2["default"].createElement(
        'table',
        { className: prefixCls + '-table', cellSpacing: '0', role: 'grid', ref: function ref(c) {
            _this2.fullTable = c;
          } },
        _react2["default"].createElement(_DateTHead2["default"], this.props),
        _react2["default"].createElement(_DateTBody2["default"], this.props)
      ),
      this.renderEvents()
    );
  };

  return DateTable;
}(_react2["default"].Component);

DateTable.propTypes = {
  localeStr: _propTypes2["default"].string
};
DateTable.defaultProps = {
  localeStr: 'zh-cn'
};
exports["default"] = DateTable;
module.exports = exports['default'];
