// uxcore-crumb/build/CrumbItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * CrumbItem Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author peijie.dpj
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2015, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var CrumbItem = function (_React$Component) {
  _inherits(CrumbItem, _React$Component);

  function CrumbItem() {
    _classCallCheck(this, CrumbItem);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  CrumbItem.prototype.render = function render() {
    var props = _objectWithoutProperties(this.props, []);

    if (props.onClick) {
      props.href = "javascript:void(0);";
    }
    return props.href ? _react2["default"].createElement(
      'a',
      props,
      props.children
    ) : _react2["default"].createElement(
      'span',
      props,
      props.children
    );
  };

  return CrumbItem;
}(_react2["default"].Component);

CrumbItem.defaultProps = {
  children: '链接'

  // http://facebook.github.io/react/docs/reusable-components.html
};CrumbItem.propTypes = {
  /**
   * @title 链接
   */
  href: _propTypes2["default"].string,
  /**
   * @title 类名
   * @veIgnore
   */
  className: _propTypes2["default"].string,
  /**
   * @title 文案
   */
  children: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].element]),
  /**
   * @title 点击事件
   * @veIgnore
   */
  onClick: _propTypes2["default"].func
};

CrumbItem.displayName = 'CrumbItem';

exports["default"] = CrumbItem;
module.exports = exports['default'];
