// uxcore-table/build/CollapsedButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _uxcoreButton = require('uxcore-button');

var _uxcoreButton2 = _interopRequireDefault(_uxcoreButton);

var _uxcoreDropdown = require('uxcore-dropdown');

var _uxcoreDropdown2 = _interopRequireDefault(_uxcoreDropdown);

var _uxcoreMenu = require('uxcore-menu');

var _uxcoreMenu2 = _interopRequireDefault(_uxcoreMenu);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _i18n = require('./i18n');

var _i18n2 = _interopRequireDefault(_i18n);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

/**
 * only consider the button style and how to collapse
 */

var CollapsedButtonItem = function CollapsedButtonItem() {};

var CollapsedButton = function (_React$Component) {
  _inherits(CollapsedButton, _React$Component);

  function CollapsedButton(props) {
    _classCallCheck(this, CollapsedButton);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      dropdownVisible: false
    };
    _this.handleDropdownVisibleChange = _this.handleDropdownVisibleChange.bind(_this);
    _this.handleMoreClick = _this.handleMoreClick.bind(_this);
    _this.refCallback = [];
    return _this;
  }

  CollapsedButton.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if (_react2["default"].Children.count(nextProps.children) < parseInt(nextProps.maxLength, 10)) {
      this.setState({
        dropdownVisible: false
      });
    }
  };

  CollapsedButton.prototype.componentDidUpdate = function componentDidUpdate() {
    var _props = this.props,
        children = _props.children,
        maxLength = _props.maxLength;

    if (this.state.dropdownVisible && parseInt(maxLength, 10) === 1 && _react2["default"].Children.count(children) > 1) {
      var dropdownDOMNode = this.dropdownInstance.getPopupDomNode();
      /* eslint-disable react/no-find-dom-node */
      var triggerDOMnode = _reactDom2["default"].findDOMNode(this.triggerInstance);
      /* eslint-enable react/no-find-dom-node */
      dropdownDOMNode.style.minWidth = (triggerDOMnode || this.triggerInstance).offsetWidth + 'px';
    }
  };

  CollapsedButton.prototype.saveRef = function saveRef(refName) {
    var me = this;
    if (me.refCallback[refName]) {
      return me.refCallback[refName];
    }
    me.refCallback[refName] = function (c) {
      me[refName] = c;
      return false;
    };
    return me.refCallback[refName];
  };

  CollapsedButton.prototype.handleDropdownVisibleChange = function handleDropdownVisibleChange(visible) {
    var me = this;
    me.setState({
      dropdownVisible: visible
    });
  };

  CollapsedButton.prototype.handleMenuItemClick = function handleMenuItemClick(action, e) {
    if (action.props.onClick) {
      action.props.onClick(e);
    }
    this.setState({
      dropdownVisible: false
    });
  };

  CollapsedButton.prototype.handleMoreClick = function handleMoreClick(e) {
    e.preventDefault();
    var me = this;
    me.setState({
      dropdownVisible: !me.state.dropdownVisible
    });
  };

  CollapsedButton.prototype.renderItem = function renderItem(item, index) {
    var me = this;
    var type = me.props.type;

    var itemProps = {
      className: 'action',
      key: index,
      onClick: item.props.disabled ? function () {} : item.props.onClick
    };
    if (type === 'button') {
      return _react2["default"].createElement(
        _uxcoreButton2["default"],
        _extends({}, itemProps, {
          type: 'outline',
          size: 'small',
          disabled: !!item.props.disabled
        }),
        item.props.children
      );
    }
    return _react2["default"].createElement(
      'a',
      _extends({}, itemProps, {
        className: (0, _classnames2["default"])({
          action: true,
          disabled: !!item.props.disabled
        })
      }),
      item.props.children
    );
  };

  CollapsedButton.prototype.renderMore = function renderMore(actions) {
    if (actions.length === 0) {
      return null;
    }
    var me = this;
    var menu = _react2["default"].createElement(
      _uxcoreMenu2["default"],
      null,
      actions.map(function (action, index) {
        return _react2["default"].createElement(
          _uxcoreMenu2["default"].Item,
          { key: index, disabled: !!action.props.disabled },
          _react2["default"].createElement(
            'a',
            { onClick: me.handleMenuItemClick.bind(me, action) },
            action.props.children
          )
        );
      })
    );

    var offsetY = me.props.type === 'button' ? -5 : -20;
    var dropdownOptions = {
      key: 'icon',
      overlay: menu,
      trigger: ['click'],
      align: {
        offset: [0, offsetY]
      },
      visible: me.state.dropdownVisible,
      overlayClassName: me.props.prefixCls + '-collapsed-button-more-dropdown',
      onVisibleChange: me.handleDropdownVisibleChange
    };
    var content = _react2["default"].createElement(
      'span',
      null,
      _i18n2["default"][me.props.locale].more,
      _react2["default"].createElement('i', {
        className: (0, _classnames2["default"])({
          'kuma-icon': true,
          'kuma-icon-triangle-down': !me.state.dropdownVisible,
          'kuma-icon-triangle-up': me.state.dropdownVisible
        })
      })
    );
    if (me.props.type === 'button') {
      return _react2["default"].createElement(
        _uxcoreDropdown2["default"],
        dropdownOptions,
        _react2["default"].createElement(
          _uxcoreButton2["default"],
          { className: 'action', type: 'outline', size: 'small' },
          content
        )
      );
    }
    return _react2["default"].createElement(
      _uxcoreDropdown2["default"],
      dropdownOptions,
      _react2["default"].createElement(
        'a',
        { className: 'action', onClick: me.handleMoreClick },
        content
      )
    );
  };

  CollapsedButton.prototype.renderHoverMenu = function renderHoverMenu() {
    var me = this;
    var _me$props = me.props,
        children = _me$props.children,
        type = _me$props.type;

    var trigger = void 0;
    var options = [];
    _react2["default"].Children.forEach(children, function (child, index) {
      if (index === 0) {
        var triggerContent = _react2["default"].createElement(
          'span',
          null,
          child.props.children,
          _react2["default"].createElement('i', {
            className: (0, _classnames2["default"])({
              'kuma-icon': true,
              'kuma-icon-triangle-down': !me.state.dropdownVisible,
              'kuma-icon-triangle-up': me.state.dropdownVisible
            })
          })
        );
        if (type === 'button') {
          trigger = _react2["default"].createElement(
            _uxcoreButton2["default"],
            {
              className: 'action',
              type: 'outline',
              size: 'small',
              ref: me.saveRef('triggerInstance')
            },
            triggerContent
          );
        } else {
          trigger = _react2["default"].createElement(
            'a',
            { className: 'action collapse-one', ref: me.saveRef('triggerInstance') },
            triggerContent
          );
        }
      }
      options.push(_react2["default"].createElement(
        _uxcoreMenu2["default"].Item,
        { key: index, disabled: !!child.props.disabled },
        _react2["default"].createElement(
          'a',
          {
            onClick: me.handleMenuItemClick.bind(me, child)
          },
          child.props.children
        )
      ));
    });
    var menu = _react2["default"].createElement(
      _uxcoreMenu2["default"],
      null,
      options
    );

    var offsetY = type === 'button' ? -31 : -44;

    var dropdownOptions = {
      key: 'icon',
      overlay: menu,
      transitionName: '',
      ref: me.saveRef('dropdownInstance'),
      trigger: ['hover'],
      align: {
        offset: [0, offsetY]
      },
      visible: me.state.dropdownVisible,
      overlayClassName: me.props.prefixCls + '-collapsed-button-more-dropdown',
      onVisibleChange: me.handleDropdownVisibleChange
    };

    return _react2["default"].createElement(
      _uxcoreDropdown2["default"],
      dropdownOptions,
      trigger
    );
  };

  CollapsedButton.prototype.render = function render() {
    var me = this;
    var _me$props2 = me.props,
        children = _me$props2.children,
        maxLength = _me$props2.maxLength;

    var buttons = [];
    var options = [];
    if (parseInt(maxLength, 10) === 1 && _react2["default"].Children.count(children) > 1) {
      return _react2["default"].createElement(
        'div',
        null,
        me.renderHoverMenu()
      );
    }
    if (_react2["default"].Children.count(children) <= parseInt(maxLength, 10)) {
      _react2["default"].Children.forEach(children, function (item, index) {
        buttons.push(me.renderItem(item, index));
      });
    } else {
      _react2["default"].Children.forEach(children, function (item, index) {
        if (index < parseInt(maxLength, 10) - 1) {
          buttons.push(me.renderItem(item, index));
        } else {
          options.push(item);
        }
      });
    }

    return _react2["default"].createElement(
      'div',
      null,
      buttons,
      me.renderMore(options)
    );
  };

  return CollapsedButton;
}(_react2["default"].Component);

CollapsedButton.displayName = 'CollapsedButton';

CollapsedButton.Item = CollapsedButtonItem;

CollapsedButton.propTypes = {
  children: _react2["default"].PropTypes.any,
  locale: _react2["default"].PropTypes.string,
  maxLength: _react2["default"].PropTypes.number,
  onClick: _react2["default"].PropTypes.func
};
CollapsedButton.defaultProps = {
  maxLength: 3,
  onClick: function onClick() {}
};

exports["default"] = CollapsedButton;
module.exports = exports['default'];
