// zent/es/steps/Steps.js
import { __assign, __extends, __rest } from "tslib";
import * as React from 'react';
import { Component } from 'react';
import Step from './components/Step';
import NumberSteps from './components/NumberSteps';
import BreadcrumbSteps from './components/BreadcrumbSteps';
var Steps = (function (_super) {
    __extends(Steps, _super);
    function Steps() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Steps.prototype.render = function () {
        var props = this.props;
        var type = props.type, children = props.children, restProps = __rest(props, ["type", "children"]);
        var typeComponentMapping = {
            number: NumberSteps,
            card: BreadcrumbSteps,
            breadcrumb: BreadcrumbSteps,
            tabs: BreadcrumbSteps,
        };
        var StepsComponent = typeComponentMapping[type];
        return (React.createElement(StepsComponent, __assign({}, restProps, { type: type }), children));
    };
    Steps.defaultProps = {
        className: '',
        prefix: 'zent',
        type: 'number',
        current: 0,
        direction: 'horizontal',
        size: 'normal',
        status: 'process',
        sequence: true,
    };
    Steps.Step = Step;
    return Steps;
}(Component));
export { Steps };
export default Steps;
