// uxcore-timeline/build/Timeline.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _TimelineItem = require('./TimelineItem');

var _TimelineItem2 = _interopRequireDefault(_TimelineItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Timeline Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author muwen.lb
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2015-2018, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var Timeline = function (_React$Component) {
  _inherits(Timeline, _React$Component);

  function Timeline(props) {
    _classCallCheck(this, Timeline);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      pending: props.pending
    };
    return _this;
  }

  Timeline.prototype.render = function render() {
    var _classnames,
        _this2 = this;

    var prefixCls = this.props.prefixCls;
    var pending = this.state.pending;
    var timelineClassName = (0, _classnames3["default"])((_classnames = {}, _defineProperty(_classnames, prefixCls, true), _defineProperty(_classnames, this.props.className, this.props.className), _classnames));
    return _react2["default"].createElement(
      'div',
      { className: timelineClassName },
      _react2["default"].createElement(
        'ul',
        { className: pending ? prefixCls + '-pending' : '' },
        _react2["default"].Children.map(this.props.children, function (ele, idx) {
          return _react2["default"].cloneElement(ele, {
            last: idx === _this2.props.children.length - 1
          });
        }),
        pending ? _react2["default"].createElement(
          _TimelineItem2["default"],
          { pending: true, color: 'gray' },
          pending
        ) : null
      )
    );
  };

  return Timeline;
}(_react2["default"].Component);

Timeline.defaultProps = {
  className: '',
  prefixCls: 'kuma-timeline'
};
Timeline.propTypes = {
  className: _propTypes2["default"].string,
  prefixCls: _propTypes2["default"].string,
  pending: _propTypes2["default"].oneOfType([_propTypes2["default"].element, _propTypes2["default"].string, _propTypes2["default"].number]),
  children: _propTypes2["default"].any
};
Timeline.Item = _TimelineItem2["default"];
Timeline.displayName = 'Timeline';
exports["default"] = Timeline;
module.exports = exports['default'];
