// uxcore-timeline/build/TimelineItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames3 = require('classnames');

var _classnames4 = _interopRequireDefault(_classnames3);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Timeline Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author muwen.lb
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2015-2018, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var TimelineItem = function (_Component) {
  _inherits(TimelineItem, _Component);

  function TimelineItem(props) {
    _classCallCheck(this, TimelineItem);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.state = {
      dot: props.dot
    };
    return _this;
  }

  TimelineItem.prototype.render = function render() {
    var _classnames, _classnames2;

    var prefixCls = this.props.prefixCls;
    var itemClassName = (0, _classnames4["default"])((_classnames = {}, _defineProperty(_classnames, prefixCls + '-item', true), _defineProperty(_classnames, prefixCls + '-item-last', this.props.last), _defineProperty(_classnames, prefixCls + '-item-pending', this.props.pending), _defineProperty(_classnames, prefixCls + '-item-dotted', this.props.dotted), _defineProperty(_classnames, prefixCls + '-item-active', this.props.active), _defineProperty(_classnames, this.props.className, this.props.className), _classnames));

    var dotClassName = (0, _classnames4["default"])((_classnames2 = {}, _defineProperty(_classnames2, prefixCls + '-item-icon', true), _defineProperty(_classnames2, prefixCls + '-item-icon-custom', this.state.dot), _classnames2));

    return _react2["default"].createElement(
      'li',
      { className: itemClassName },
      _react2["default"].createElement('div', {
        className: prefixCls + '-item-line',
        style: {
          borderColor: this.props.color || null
        }
      }),
      _react2["default"].createElement(
        'div',
        {
          className: dotClassName,
          style: {
            backgroundColor: this.props.color || null
          }
        },
        this.state.dot
      ),
      this.props.title ? _react2["default"].createElement(
        'div',
        {
          className: prefixCls + '-item-title',
          style: {
            color: this.props.color || null
          }
        },
        this.props.title
      ) : null,
      _react2["default"].createElement(
        'div',
        { className: prefixCls + '-item-content' },
        this.props.children
      )
    );
  };

  return TimelineItem;
}(_react.Component);

TimelineItem.displayName = 'Timeline';
TimelineItem.defaultProps = {
  pending: false,
  prefixCls: 'kuma-timeline',
  className: '', // 额外的className
  last: false, // 是否最后一个
  color: '', // 自定义颜色
  dotted: false, // 自定义实线虚线
  active: false, // 进行中的时间节点
  title: ''
};
TimelineItem.propTypes = {
  pending: _propTypes2["default"].bool,
  prefixCls: _propTypes2["default"].string,
  className: _propTypes2["default"].string,
  last: _propTypes2["default"].bool,
  color: _propTypes2["default"].string,
  dotted: _propTypes2["default"].bool,
  active: _propTypes2["default"].bool,
  title: _propTypes2["default"].oneOfType([_propTypes2["default"].element, _propTypes2["default"].string]),
  dot: _propTypes2["default"].oneOfType([_propTypes2["default"].element, _propTypes2["default"].string]),
  children: _propTypes2["default"].any
};
exports["default"] = TimelineItem;
module.exports = exports['default'];
